{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "509748fa",
   "metadata": {},
   "source": [
    "# Assignment 10: Breast cancer gene expression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f2e0dac",
   "metadata": {},
   "source": [
    "## Modules and packages import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "705c0ef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import zscore, ttest_ind\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import gseapy as gp\n",
    "import numpy as np\n",
    "from gseapy import barplot, dotplot, heatmap\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ef696ff",
   "metadata": {},
   "source": [
    "## Dataset import and preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9f74ddfb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID_REF</th>\n",
       "      <th>100002</th>\n",
       "      <th>100003</th>\n",
       "      <th>100027</th>\n",
       "      <th>100036</th>\n",
       "      <th>100037</th>\n",
       "      <th>100039</th>\n",
       "      <th>100044</th>\n",
       "      <th>100045</th>\n",
       "      <th>100051</th>\n",
       "      <th>100052</th>\n",
       "      <th>...</th>\n",
       "      <th>10713472</th>\n",
       "      <th>10713880</th>\n",
       "      <th>10713947</th>\n",
       "      <th>10714494</th>\n",
       "      <th>10715276</th>\n",
       "      <th>10715317</th>\n",
       "      <th>10715552</th>\n",
       "      <th>10715690</th>\n",
       "      <th>10715781</th>\n",
       "      <th>10715908</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM290763</th>\n",
       "      <td>13.422506</td>\n",
       "      <td>7.350945</td>\n",
       "      <td>8.178899</td>\n",
       "      <td>11.923099</td>\n",
       "      <td>12.809180</td>\n",
       "      <td>9.125464</td>\n",
       "      <td>10.624943</td>\n",
       "      <td>8.099086</td>\n",
       "      <td>8.639951</td>\n",
       "      <td>8.903277</td>\n",
       "      <td>...</td>\n",
       "      <td>8.291419</td>\n",
       "      <td>10.441818</td>\n",
       "      <td>7.613034</td>\n",
       "      <td>12.118979</td>\n",
       "      <td>7.654639</td>\n",
       "      <td>9.722318</td>\n",
       "      <td>16.142613</td>\n",
       "      <td>10.677725</td>\n",
       "      <td>12.001682</td>\n",
       "      <td>7.370101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290764</th>\n",
       "      <td>13.780101</td>\n",
       "      <td>6.499586</td>\n",
       "      <td>7.575710</td>\n",
       "      <td>11.484546</td>\n",
       "      <td>12.754843</td>\n",
       "      <td>7.117874</td>\n",
       "      <td>7.789392</td>\n",
       "      <td>9.247552</td>\n",
       "      <td>7.760204</td>\n",
       "      <td>7.417160</td>\n",
       "      <td>...</td>\n",
       "      <td>10.310325</td>\n",
       "      <td>9.530774</td>\n",
       "      <td>7.052968</td>\n",
       "      <td>12.213755</td>\n",
       "      <td>8.261598</td>\n",
       "      <td>10.103332</td>\n",
       "      <td>16.122971</td>\n",
       "      <td>11.534957</td>\n",
       "      <td>11.298948</td>\n",
       "      <td>9.204856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290765</th>\n",
       "      <td>15.062165</td>\n",
       "      <td>9.109690</td>\n",
       "      <td>8.880089</td>\n",
       "      <td>11.822615</td>\n",
       "      <td>13.807391</td>\n",
       "      <td>9.777046</td>\n",
       "      <td>8.087078</td>\n",
       "      <td>9.007773</td>\n",
       "      <td>7.731535</td>\n",
       "      <td>7.699856</td>\n",
       "      <td>...</td>\n",
       "      <td>9.584304</td>\n",
       "      <td>7.438050</td>\n",
       "      <td>8.530162</td>\n",
       "      <td>12.144980</td>\n",
       "      <td>6.978764</td>\n",
       "      <td>9.143822</td>\n",
       "      <td>16.666508</td>\n",
       "      <td>11.459627</td>\n",
       "      <td>9.507795</td>\n",
       "      <td>10.344573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290766</th>\n",
       "      <td>13.148520</td>\n",
       "      <td>8.520583</td>\n",
       "      <td>7.696917</td>\n",
       "      <td>12.535856</td>\n",
       "      <td>11.997358</td>\n",
       "      <td>7.766018</td>\n",
       "      <td>9.022442</td>\n",
       "      <td>9.180375</td>\n",
       "      <td>7.645393</td>\n",
       "      <td>8.527599</td>\n",
       "      <td>...</td>\n",
       "      <td>8.543886</td>\n",
       "      <td>10.103882</td>\n",
       "      <td>8.031094</td>\n",
       "      <td>11.855796</td>\n",
       "      <td>8.562418</td>\n",
       "      <td>10.266450</td>\n",
       "      <td>16.183008</td>\n",
       "      <td>10.366325</td>\n",
       "      <td>12.302811</td>\n",
       "      <td>9.927861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290767</th>\n",
       "      <td>14.132942</td>\n",
       "      <td>8.594196</td>\n",
       "      <td>7.405297</td>\n",
       "      <td>10.624943</td>\n",
       "      <td>13.604827</td>\n",
       "      <td>10.538310</td>\n",
       "      <td>7.541922</td>\n",
       "      <td>7.171625</td>\n",
       "      <td>6.779122</td>\n",
       "      <td>8.366277</td>\n",
       "      <td>...</td>\n",
       "      <td>10.021877</td>\n",
       "      <td>9.469259</td>\n",
       "      <td>8.566389</td>\n",
       "      <td>12.210368</td>\n",
       "      <td>5.284179</td>\n",
       "      <td>10.423989</td>\n",
       "      <td>17.409784</td>\n",
       "      <td>11.934193</td>\n",
       "      <td>10.640237</td>\n",
       "      <td>9.766645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290820</th>\n",
       "      <td>14.835224</td>\n",
       "      <td>8.339718</td>\n",
       "      <td>7.695143</td>\n",
       "      <td>10.665347</td>\n",
       "      <td>13.814970</td>\n",
       "      <td>9.456169</td>\n",
       "      <td>8.363736</td>\n",
       "      <td>7.983704</td>\n",
       "      <td>8.110529</td>\n",
       "      <td>8.942591</td>\n",
       "      <td>...</td>\n",
       "      <td>9.763001</td>\n",
       "      <td>8.894234</td>\n",
       "      <td>8.433577</td>\n",
       "      <td>11.796304</td>\n",
       "      <td>9.454508</td>\n",
       "      <td>9.046134</td>\n",
       "      <td>16.881217</td>\n",
       "      <td>11.521516</td>\n",
       "      <td>9.433552</td>\n",
       "      <td>9.370892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290821</th>\n",
       "      <td>14.814234</td>\n",
       "      <td>8.017298</td>\n",
       "      <td>8.801931</td>\n",
       "      <td>10.104682</td>\n",
       "      <td>13.650664</td>\n",
       "      <td>10.642495</td>\n",
       "      <td>6.845055</td>\n",
       "      <td>7.938469</td>\n",
       "      <td>9.451076</td>\n",
       "      <td>7.174347</td>\n",
       "      <td>...</td>\n",
       "      <td>10.172256</td>\n",
       "      <td>7.865130</td>\n",
       "      <td>8.424459</td>\n",
       "      <td>12.164485</td>\n",
       "      <td>7.249902</td>\n",
       "      <td>9.688956</td>\n",
       "      <td>17.149760</td>\n",
       "      <td>11.740188</td>\n",
       "      <td>9.838539</td>\n",
       "      <td>9.302903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290822</th>\n",
       "      <td>14.397616</td>\n",
       "      <td>8.558820</td>\n",
       "      <td>9.520554</td>\n",
       "      <td>9.866795</td>\n",
       "      <td>13.529753</td>\n",
       "      <td>9.769843</td>\n",
       "      <td>7.891491</td>\n",
       "      <td>7.473685</td>\n",
       "      <td>7.700255</td>\n",
       "      <td>9.204245</td>\n",
       "      <td>...</td>\n",
       "      <td>10.514198</td>\n",
       "      <td>9.168202</td>\n",
       "      <td>8.162785</td>\n",
       "      <td>11.997358</td>\n",
       "      <td>7.599954</td>\n",
       "      <td>10.852385</td>\n",
       "      <td>16.965637</td>\n",
       "      <td>11.304308</td>\n",
       "      <td>9.778815</td>\n",
       "      <td>11.096939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290823</th>\n",
       "      <td>15.268965</td>\n",
       "      <td>7.574579</td>\n",
       "      <td>7.756522</td>\n",
       "      <td>9.688699</td>\n",
       "      <td>14.372324</td>\n",
       "      <td>10.985775</td>\n",
       "      <td>7.384741</td>\n",
       "      <td>6.822938</td>\n",
       "      <td>7.771283</td>\n",
       "      <td>7.876163</td>\n",
       "      <td>...</td>\n",
       "      <td>9.766561</td>\n",
       "      <td>8.882471</td>\n",
       "      <td>8.695544</td>\n",
       "      <td>12.061112</td>\n",
       "      <td>7.594228</td>\n",
       "      <td>9.535737</td>\n",
       "      <td>16.441746</td>\n",
       "      <td>11.551248</td>\n",
       "      <td>8.290506</td>\n",
       "      <td>9.100522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290824</th>\n",
       "      <td>15.001767</td>\n",
       "      <td>8.250282</td>\n",
       "      <td>8.652816</td>\n",
       "      <td>10.023413</td>\n",
       "      <td>14.179916</td>\n",
       "      <td>10.127081</td>\n",
       "      <td>7.878391</td>\n",
       "      <td>7.776379</td>\n",
       "      <td>9.301350</td>\n",
       "      <td>7.241005</td>\n",
       "      <td>...</td>\n",
       "      <td>9.190938</td>\n",
       "      <td>8.503289</td>\n",
       "      <td>7.481404</td>\n",
       "      <td>11.845281</td>\n",
       "      <td>7.408495</td>\n",
       "      <td>9.598420</td>\n",
       "      <td>16.845547</td>\n",
       "      <td>12.147251</td>\n",
       "      <td>9.700231</td>\n",
       "      <td>11.599191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 32878 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID_REF      100002    100003    100027     100036     100037     100039    \\\n",
       "GSM290763  13.422506  7.350945  8.178899  11.923099  12.809180   9.125464   \n",
       "GSM290764  13.780101  6.499586  7.575710  11.484546  12.754843   7.117874   \n",
       "GSM290765  15.062165  9.109690  8.880089  11.822615  13.807391   9.777046   \n",
       "GSM290766  13.148520  8.520583  7.696917  12.535856  11.997358   7.766018   \n",
       "GSM290767  14.132942  8.594196  7.405297  10.624943  13.604827  10.538310   \n",
       "...              ...       ...       ...        ...        ...        ...   \n",
       "GSM290820  14.835224  8.339718  7.695143  10.665347  13.814970   9.456169   \n",
       "GSM290821  14.814234  8.017298  8.801931  10.104682  13.650664  10.642495   \n",
       "GSM290822  14.397616  8.558820  9.520554   9.866795  13.529753   9.769843   \n",
       "GSM290823  15.268965  7.574579  7.756522   9.688699  14.372324  10.985775   \n",
       "GSM290824  15.001767  8.250282  8.652816  10.023413  14.179916  10.127081   \n",
       "\n",
       "ID_REF      100044    100045    100051    100052    ...   10713472   10713880  \\\n",
       "GSM290763  10.624943  8.099086  8.639951  8.903277  ...   8.291419  10.441818   \n",
       "GSM290764   7.789392  9.247552  7.760204  7.417160  ...  10.310325   9.530774   \n",
       "GSM290765   8.087078  9.007773  7.731535  7.699856  ...   9.584304   7.438050   \n",
       "GSM290766   9.022442  9.180375  7.645393  8.527599  ...   8.543886  10.103882   \n",
       "GSM290767   7.541922  7.171625  6.779122  8.366277  ...  10.021877   9.469259   \n",
       "...              ...       ...       ...       ...  ...        ...        ...   \n",
       "GSM290820   8.363736  7.983704  8.110529  8.942591  ...   9.763001   8.894234   \n",
       "GSM290821   6.845055  7.938469  9.451076  7.174347  ...  10.172256   7.865130   \n",
       "GSM290822   7.891491  7.473685  7.700255  9.204245  ...  10.514198   9.168202   \n",
       "GSM290823   7.384741  6.822938  7.771283  7.876163  ...   9.766561   8.882471   \n",
       "GSM290824   7.878391  7.776379  9.301350  7.241005  ...   9.190938   8.503289   \n",
       "\n",
       "ID_REF     10713947   10714494  10715276   10715317   10715552   10715690  \\\n",
       "GSM290763  7.613034  12.118979  7.654639   9.722318  16.142613  10.677725   \n",
       "GSM290764  7.052968  12.213755  8.261598  10.103332  16.122971  11.534957   \n",
       "GSM290765  8.530162  12.144980  6.978764   9.143822  16.666508  11.459627   \n",
       "GSM290766  8.031094  11.855796  8.562418  10.266450  16.183008  10.366325   \n",
       "GSM290767  8.566389  12.210368  5.284179  10.423989  17.409784  11.934193   \n",
       "...             ...        ...       ...        ...        ...        ...   \n",
       "GSM290820  8.433577  11.796304  9.454508   9.046134  16.881217  11.521516   \n",
       "GSM290821  8.424459  12.164485  7.249902   9.688956  17.149760  11.740188   \n",
       "GSM290822  8.162785  11.997358  7.599954  10.852385  16.965637  11.304308   \n",
       "GSM290823  8.695544  12.061112  7.594228   9.535737  16.441746  11.551248   \n",
       "GSM290824  7.481404  11.845281  7.408495   9.598420  16.845547  12.147251   \n",
       "\n",
       "ID_REF      10715781   10715908  \n",
       "GSM290763  12.001682   7.370101  \n",
       "GSM290764  11.298948   9.204856  \n",
       "GSM290765   9.507795  10.344573  \n",
       "GSM290766  12.302811   9.927861  \n",
       "GSM290767  10.640237   9.766645  \n",
       "...              ...        ...  \n",
       "GSM290820   9.433552   9.370892  \n",
       "GSM290821   9.838539   9.302903  \n",
       "GSM290822   9.778815  11.096939  \n",
       "GSM290823   8.290506   9.100522  \n",
       "GSM290824   9.700231  11.599191  \n",
       "\n",
       "[62 rows x 32878 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('dataset_breast_cancer.txt', sep='\\t', index_col='ID_REF').transpose()\n",
    "df  \n",
    "#We import the complete dataset, from which we will sample just data referring to healthy or breast cancer people"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f52444c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID_REF</th>\n",
       "      <th>100002</th>\n",
       "      <th>100003</th>\n",
       "      <th>100027</th>\n",
       "      <th>100036</th>\n",
       "      <th>100037</th>\n",
       "      <th>100039</th>\n",
       "      <th>100044</th>\n",
       "      <th>100045</th>\n",
       "      <th>100051</th>\n",
       "      <th>100052</th>\n",
       "      <th>...</th>\n",
       "      <th>10713472</th>\n",
       "      <th>10713880</th>\n",
       "      <th>10713947</th>\n",
       "      <th>10714494</th>\n",
       "      <th>10715276</th>\n",
       "      <th>10715317</th>\n",
       "      <th>10715552</th>\n",
       "      <th>10715690</th>\n",
       "      <th>10715781</th>\n",
       "      <th>10715908</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM290816</th>\n",
       "      <td>13.710818</td>\n",
       "      <td>6.892902</td>\n",
       "      <td>8.711001</td>\n",
       "      <td>11.998301</td>\n",
       "      <td>12.888501</td>\n",
       "      <td>8.875158</td>\n",
       "      <td>7.950999</td>\n",
       "      <td>7.536447</td>\n",
       "      <td>7.547635</td>\n",
       "      <td>8.083784</td>\n",
       "      <td>...</td>\n",
       "      <td>8.185244</td>\n",
       "      <td>7.615529</td>\n",
       "      <td>8.865868</td>\n",
       "      <td>11.005292</td>\n",
       "      <td>7.691663</td>\n",
       "      <td>9.456169</td>\n",
       "      <td>16.624669</td>\n",
       "      <td>11.643177</td>\n",
       "      <td>10.737178</td>\n",
       "      <td>9.864491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290817</th>\n",
       "      <td>14.905733</td>\n",
       "      <td>5.663721</td>\n",
       "      <td>7.370765</td>\n",
       "      <td>10.746211</td>\n",
       "      <td>13.683590</td>\n",
       "      <td>10.629111</td>\n",
       "      <td>8.462847</td>\n",
       "      <td>8.223309</td>\n",
       "      <td>7.460839</td>\n",
       "      <td>8.047821</td>\n",
       "      <td>...</td>\n",
       "      <td>8.401635</td>\n",
       "      <td>9.332863</td>\n",
       "      <td>7.781298</td>\n",
       "      <td>11.566280</td>\n",
       "      <td>8.103638</td>\n",
       "      <td>9.060828</td>\n",
       "      <td>16.219681</td>\n",
       "      <td>11.168915</td>\n",
       "      <td>9.639253</td>\n",
       "      <td>8.276696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290818</th>\n",
       "      <td>13.093811</td>\n",
       "      <td>8.082162</td>\n",
       "      <td>7.616559</td>\n",
       "      <td>11.958278</td>\n",
       "      <td>12.248943</td>\n",
       "      <td>7.992173</td>\n",
       "      <td>7.788214</td>\n",
       "      <td>7.604562</td>\n",
       "      <td>7.739810</td>\n",
       "      <td>7.399937</td>\n",
       "      <td>...</td>\n",
       "      <td>9.069586</td>\n",
       "      <td>9.288055</td>\n",
       "      <td>7.528501</td>\n",
       "      <td>11.753377</td>\n",
       "      <td>7.085317</td>\n",
       "      <td>10.170410</td>\n",
       "      <td>16.219681</td>\n",
       "      <td>11.690165</td>\n",
       "      <td>10.567470</td>\n",
       "      <td>8.454736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290819</th>\n",
       "      <td>14.740331</td>\n",
       "      <td>8.918132</td>\n",
       "      <td>7.724920</td>\n",
       "      <td>9.851807</td>\n",
       "      <td>14.121267</td>\n",
       "      <td>9.555641</td>\n",
       "      <td>7.718085</td>\n",
       "      <td>7.276954</td>\n",
       "      <td>7.883154</td>\n",
       "      <td>7.812956</td>\n",
       "      <td>...</td>\n",
       "      <td>10.182163</td>\n",
       "      <td>9.102335</td>\n",
       "      <td>7.011307</td>\n",
       "      <td>11.946573</td>\n",
       "      <td>8.608208</td>\n",
       "      <td>10.328690</td>\n",
       "      <td>16.949791</td>\n",
       "      <td>11.823872</td>\n",
       "      <td>9.191861</td>\n",
       "      <td>8.743873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290820</th>\n",
       "      <td>14.835224</td>\n",
       "      <td>8.339718</td>\n",
       "      <td>7.695143</td>\n",
       "      <td>10.665347</td>\n",
       "      <td>13.814970</td>\n",
       "      <td>9.456169</td>\n",
       "      <td>8.363736</td>\n",
       "      <td>7.983704</td>\n",
       "      <td>8.110529</td>\n",
       "      <td>8.942591</td>\n",
       "      <td>...</td>\n",
       "      <td>9.763001</td>\n",
       "      <td>8.894234</td>\n",
       "      <td>8.433577</td>\n",
       "      <td>11.796304</td>\n",
       "      <td>9.454508</td>\n",
       "      <td>9.046134</td>\n",
       "      <td>16.881217</td>\n",
       "      <td>11.521516</td>\n",
       "      <td>9.433552</td>\n",
       "      <td>9.370892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290821</th>\n",
       "      <td>14.814234</td>\n",
       "      <td>8.017298</td>\n",
       "      <td>8.801931</td>\n",
       "      <td>10.104682</td>\n",
       "      <td>13.650664</td>\n",
       "      <td>10.642495</td>\n",
       "      <td>6.845055</td>\n",
       "      <td>7.938469</td>\n",
       "      <td>9.451076</td>\n",
       "      <td>7.174347</td>\n",
       "      <td>...</td>\n",
       "      <td>10.172256</td>\n",
       "      <td>7.865130</td>\n",
       "      <td>8.424459</td>\n",
       "      <td>12.164485</td>\n",
       "      <td>7.249902</td>\n",
       "      <td>9.688956</td>\n",
       "      <td>17.149760</td>\n",
       "      <td>11.740188</td>\n",
       "      <td>9.838539</td>\n",
       "      <td>9.302903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290822</th>\n",
       "      <td>14.397616</td>\n",
       "      <td>8.558820</td>\n",
       "      <td>9.520554</td>\n",
       "      <td>9.866795</td>\n",
       "      <td>13.529753</td>\n",
       "      <td>9.769843</td>\n",
       "      <td>7.891491</td>\n",
       "      <td>7.473685</td>\n",
       "      <td>7.700255</td>\n",
       "      <td>9.204245</td>\n",
       "      <td>...</td>\n",
       "      <td>10.514198</td>\n",
       "      <td>9.168202</td>\n",
       "      <td>8.162785</td>\n",
       "      <td>11.997358</td>\n",
       "      <td>7.599954</td>\n",
       "      <td>10.852385</td>\n",
       "      <td>16.965637</td>\n",
       "      <td>11.304308</td>\n",
       "      <td>9.778815</td>\n",
       "      <td>11.096939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290823</th>\n",
       "      <td>15.268965</td>\n",
       "      <td>7.574579</td>\n",
       "      <td>7.756522</td>\n",
       "      <td>9.688699</td>\n",
       "      <td>14.372324</td>\n",
       "      <td>10.985775</td>\n",
       "      <td>7.384741</td>\n",
       "      <td>6.822938</td>\n",
       "      <td>7.771283</td>\n",
       "      <td>7.876163</td>\n",
       "      <td>...</td>\n",
       "      <td>9.766561</td>\n",
       "      <td>8.882471</td>\n",
       "      <td>8.695544</td>\n",
       "      <td>12.061112</td>\n",
       "      <td>7.594228</td>\n",
       "      <td>9.535737</td>\n",
       "      <td>16.441746</td>\n",
       "      <td>11.551248</td>\n",
       "      <td>8.290506</td>\n",
       "      <td>9.100522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290824</th>\n",
       "      <td>15.001767</td>\n",
       "      <td>8.250282</td>\n",
       "      <td>8.652816</td>\n",
       "      <td>10.023413</td>\n",
       "      <td>14.179916</td>\n",
       "      <td>10.127081</td>\n",
       "      <td>7.878391</td>\n",
       "      <td>7.776379</td>\n",
       "      <td>9.301350</td>\n",
       "      <td>7.241005</td>\n",
       "      <td>...</td>\n",
       "      <td>9.190938</td>\n",
       "      <td>8.503289</td>\n",
       "      <td>7.481404</td>\n",
       "      <td>11.845281</td>\n",
       "      <td>7.408495</td>\n",
       "      <td>9.598420</td>\n",
       "      <td>16.845547</td>\n",
       "      <td>12.147251</td>\n",
       "      <td>9.700231</td>\n",
       "      <td>11.599191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 32878 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID_REF      100002    100003    100027     100036     100037     100039    \\\n",
       "GSM290816  13.710818  6.892902  8.711001  11.998301  12.888501   8.875158   \n",
       "GSM290817  14.905733  5.663721  7.370765  10.746211  13.683590  10.629111   \n",
       "GSM290818  13.093811  8.082162  7.616559  11.958278  12.248943   7.992173   \n",
       "GSM290819  14.740331  8.918132  7.724920   9.851807  14.121267   9.555641   \n",
       "GSM290820  14.835224  8.339718  7.695143  10.665347  13.814970   9.456169   \n",
       "GSM290821  14.814234  8.017298  8.801931  10.104682  13.650664  10.642495   \n",
       "GSM290822  14.397616  8.558820  9.520554   9.866795  13.529753   9.769843   \n",
       "GSM290823  15.268965  7.574579  7.756522   9.688699  14.372324  10.985775   \n",
       "GSM290824  15.001767  8.250282  8.652816  10.023413  14.179916  10.127081   \n",
       "\n",
       "ID_REF     100044    100045    100051    100052    ...   10713472  10713880  \\\n",
       "GSM290816  7.950999  7.536447  7.547635  8.083784  ...   8.185244  7.615529   \n",
       "GSM290817  8.462847  8.223309  7.460839  8.047821  ...   8.401635  9.332863   \n",
       "GSM290818  7.788214  7.604562  7.739810  7.399937  ...   9.069586  9.288055   \n",
       "GSM290819  7.718085  7.276954  7.883154  7.812956  ...  10.182163  9.102335   \n",
       "GSM290820  8.363736  7.983704  8.110529  8.942591  ...   9.763001  8.894234   \n",
       "GSM290821  6.845055  7.938469  9.451076  7.174347  ...  10.172256  7.865130   \n",
       "GSM290822  7.891491  7.473685  7.700255  9.204245  ...  10.514198  9.168202   \n",
       "GSM290823  7.384741  6.822938  7.771283  7.876163  ...   9.766561  8.882471   \n",
       "GSM290824  7.878391  7.776379  9.301350  7.241005  ...   9.190938  8.503289   \n",
       "\n",
       "ID_REF     10713947   10714494  10715276   10715317   10715552   10715690  \\\n",
       "GSM290816  8.865868  11.005292  7.691663   9.456169  16.624669  11.643177   \n",
       "GSM290817  7.781298  11.566280  8.103638   9.060828  16.219681  11.168915   \n",
       "GSM290818  7.528501  11.753377  7.085317  10.170410  16.219681  11.690165   \n",
       "GSM290819  7.011307  11.946573  8.608208  10.328690  16.949791  11.823872   \n",
       "GSM290820  8.433577  11.796304  9.454508   9.046134  16.881217  11.521516   \n",
       "GSM290821  8.424459  12.164485  7.249902   9.688956  17.149760  11.740188   \n",
       "GSM290822  8.162785  11.997358  7.599954  10.852385  16.965637  11.304308   \n",
       "GSM290823  8.695544  12.061112  7.594228   9.535737  16.441746  11.551248   \n",
       "GSM290824  7.481404  11.845281  7.408495   9.598420  16.845547  12.147251   \n",
       "\n",
       "ID_REF      10715781   10715908  \n",
       "GSM290816  10.737178   9.864491  \n",
       "GSM290817   9.639253   8.276696  \n",
       "GSM290818  10.567470   8.454736  \n",
       "GSM290819   9.191861   8.743873  \n",
       "GSM290820   9.433552   9.370892  \n",
       "GSM290821   9.838539   9.302903  \n",
       "GSM290822   9.778815  11.096939  \n",
       "GSM290823   8.290506   9.100522  \n",
       "GSM290824   9.700231  11.599191  \n",
       "\n",
       "[9 rows x 32878 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_health = df.iloc[-9:,:]\n",
    "df_health  #Dataframe with just healthy people"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "816bf8df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID_REF</th>\n",
       "      <th>100002</th>\n",
       "      <th>100003</th>\n",
       "      <th>100027</th>\n",
       "      <th>100036</th>\n",
       "      <th>100037</th>\n",
       "      <th>100039</th>\n",
       "      <th>100044</th>\n",
       "      <th>100045</th>\n",
       "      <th>100051</th>\n",
       "      <th>100052</th>\n",
       "      <th>...</th>\n",
       "      <th>10713472</th>\n",
       "      <th>10713880</th>\n",
       "      <th>10713947</th>\n",
       "      <th>10714494</th>\n",
       "      <th>10715276</th>\n",
       "      <th>10715317</th>\n",
       "      <th>10715552</th>\n",
       "      <th>10715690</th>\n",
       "      <th>10715781</th>\n",
       "      <th>10715908</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM290792</th>\n",
       "      <td>13.783817</td>\n",
       "      <td>7.894258</td>\n",
       "      <td>7.836896</td>\n",
       "      <td>11.556448</td>\n",
       "      <td>13.296613</td>\n",
       "      <td>9.808581</td>\n",
       "      <td>8.411719</td>\n",
       "      <td>7.831912</td>\n",
       "      <td>7.760311</td>\n",
       "      <td>8.393730</td>\n",
       "      <td>...</td>\n",
       "      <td>9.716142</td>\n",
       "      <td>9.561976</td>\n",
       "      <td>8.847066</td>\n",
       "      <td>12.083454</td>\n",
       "      <td>7.325629</td>\n",
       "      <td>10.257842</td>\n",
       "      <td>16.975637</td>\n",
       "      <td>11.864794</td>\n",
       "      <td>10.520180</td>\n",
       "      <td>10.913469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290793</th>\n",
       "      <td>14.494117</td>\n",
       "      <td>8.974201</td>\n",
       "      <td>8.454090</td>\n",
       "      <td>9.939125</td>\n",
       "      <td>13.804273</td>\n",
       "      <td>10.497013</td>\n",
       "      <td>7.647509</td>\n",
       "      <td>8.688864</td>\n",
       "      <td>7.854128</td>\n",
       "      <td>7.523967</td>\n",
       "      <td>...</td>\n",
       "      <td>8.596880</td>\n",
       "      <td>8.743873</td>\n",
       "      <td>7.884095</td>\n",
       "      <td>12.179391</td>\n",
       "      <td>7.986777</td>\n",
       "      <td>9.755269</td>\n",
       "      <td>16.630226</td>\n",
       "      <td>11.887620</td>\n",
       "      <td>9.655440</td>\n",
       "      <td>9.080849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290794</th>\n",
       "      <td>13.665953</td>\n",
       "      <td>7.501625</td>\n",
       "      <td>7.278090</td>\n",
       "      <td>9.881257</td>\n",
       "      <td>13.383527</td>\n",
       "      <td>9.261186</td>\n",
       "      <td>8.182199</td>\n",
       "      <td>8.105078</td>\n",
       "      <td>7.707222</td>\n",
       "      <td>8.101607</td>\n",
       "      <td>...</td>\n",
       "      <td>8.478933</td>\n",
       "      <td>7.813959</td>\n",
       "      <td>8.957044</td>\n",
       "      <td>11.905264</td>\n",
       "      <td>7.278730</td>\n",
       "      <td>10.143787</td>\n",
       "      <td>16.378968</td>\n",
       "      <td>12.334644</td>\n",
       "      <td>10.128955</td>\n",
       "      <td>11.447662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290795</th>\n",
       "      <td>14.300739</td>\n",
       "      <td>7.200073</td>\n",
       "      <td>10.258611</td>\n",
       "      <td>10.711247</td>\n",
       "      <td>13.520033</td>\n",
       "      <td>10.150032</td>\n",
       "      <td>9.329694</td>\n",
       "      <td>7.793425</td>\n",
       "      <td>7.616146</td>\n",
       "      <td>6.122135</td>\n",
       "      <td>...</td>\n",
       "      <td>10.312558</td>\n",
       "      <td>7.819357</td>\n",
       "      <td>8.585799</td>\n",
       "      <td>12.238435</td>\n",
       "      <td>7.469507</td>\n",
       "      <td>10.289238</td>\n",
       "      <td>17.075447</td>\n",
       "      <td>11.750553</td>\n",
       "      <td>9.621316</td>\n",
       "      <td>10.624943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290796</th>\n",
       "      <td>13.983886</td>\n",
       "      <td>7.633821</td>\n",
       "      <td>7.134759</td>\n",
       "      <td>10.251259</td>\n",
       "      <td>13.151941</td>\n",
       "      <td>8.837736</td>\n",
       "      <td>7.437105</td>\n",
       "      <td>7.650994</td>\n",
       "      <td>8.994013</td>\n",
       "      <td>8.454090</td>\n",
       "      <td>...</td>\n",
       "      <td>8.848244</td>\n",
       "      <td>8.391254</td>\n",
       "      <td>7.881447</td>\n",
       "      <td>12.438635</td>\n",
       "      <td>8.713046</td>\n",
       "      <td>10.296148</td>\n",
       "      <td>16.868630</td>\n",
       "      <td>11.692572</td>\n",
       "      <td>10.847043</td>\n",
       "      <td>9.937312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290797</th>\n",
       "      <td>14.278781</td>\n",
       "      <td>7.597492</td>\n",
       "      <td>6.972344</td>\n",
       "      <td>11.385409</td>\n",
       "      <td>13.952971</td>\n",
       "      <td>9.372846</td>\n",
       "      <td>7.572322</td>\n",
       "      <td>8.198980</td>\n",
       "      <td>5.796469</td>\n",
       "      <td>8.626830</td>\n",
       "      <td>...</td>\n",
       "      <td>10.169555</td>\n",
       "      <td>8.859554</td>\n",
       "      <td>10.351108</td>\n",
       "      <td>11.777696</td>\n",
       "      <td>8.004270</td>\n",
       "      <td>9.927231</td>\n",
       "      <td>16.602697</td>\n",
       "      <td>11.770852</td>\n",
       "      <td>10.529190</td>\n",
       "      <td>8.629565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290798</th>\n",
       "      <td>14.484616</td>\n",
       "      <td>7.544880</td>\n",
       "      <td>8.021632</td>\n",
       "      <td>11.025483</td>\n",
       "      <td>14.027703</td>\n",
       "      <td>10.217827</td>\n",
       "      <td>8.433729</td>\n",
       "      <td>8.415782</td>\n",
       "      <td>7.821662</td>\n",
       "      <td>8.449017</td>\n",
       "      <td>...</td>\n",
       "      <td>8.715933</td>\n",
       "      <td>7.808374</td>\n",
       "      <td>7.967301</td>\n",
       "      <td>11.454031</td>\n",
       "      <td>8.248335</td>\n",
       "      <td>10.926243</td>\n",
       "      <td>16.383979</td>\n",
       "      <td>11.793121</td>\n",
       "      <td>10.488346</td>\n",
       "      <td>11.264579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290799</th>\n",
       "      <td>14.269023</td>\n",
       "      <td>7.886140</td>\n",
       "      <td>8.680602</td>\n",
       "      <td>9.983204</td>\n",
       "      <td>13.500088</td>\n",
       "      <td>9.915698</td>\n",
       "      <td>7.422121</td>\n",
       "      <td>8.232527</td>\n",
       "      <td>6.105773</td>\n",
       "      <td>8.972284</td>\n",
       "      <td>...</td>\n",
       "      <td>9.650487</td>\n",
       "      <td>9.119511</td>\n",
       "      <td>8.938449</td>\n",
       "      <td>11.914243</td>\n",
       "      <td>7.232726</td>\n",
       "      <td>10.346964</td>\n",
       "      <td>17.047605</td>\n",
       "      <td>11.785119</td>\n",
       "      <td>10.357344</td>\n",
       "      <td>9.754022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290800</th>\n",
       "      <td>13.958387</td>\n",
       "      <td>7.377937</td>\n",
       "      <td>8.281709</td>\n",
       "      <td>11.062754</td>\n",
       "      <td>13.511526</td>\n",
       "      <td>9.133993</td>\n",
       "      <td>8.156851</td>\n",
       "      <td>9.386527</td>\n",
       "      <td>7.196282</td>\n",
       "      <td>7.830803</td>\n",
       "      <td>...</td>\n",
       "      <td>10.213736</td>\n",
       "      <td>8.658698</td>\n",
       "      <td>9.445913</td>\n",
       "      <td>12.148274</td>\n",
       "      <td>8.022025</td>\n",
       "      <td>9.926582</td>\n",
       "      <td>16.862602</td>\n",
       "      <td>11.528524</td>\n",
       "      <td>10.634424</td>\n",
       "      <td>11.328658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290807</th>\n",
       "      <td>14.731631</td>\n",
       "      <td>7.292684</td>\n",
       "      <td>7.531574</td>\n",
       "      <td>9.735370</td>\n",
       "      <td>13.945415</td>\n",
       "      <td>10.044735</td>\n",
       "      <td>7.083466</td>\n",
       "      <td>8.140264</td>\n",
       "      <td>7.210502</td>\n",
       "      <td>7.964978</td>\n",
       "      <td>...</td>\n",
       "      <td>10.321017</td>\n",
       "      <td>8.428890</td>\n",
       "      <td>7.591398</td>\n",
       "      <td>12.150382</td>\n",
       "      <td>7.318477</td>\n",
       "      <td>10.152613</td>\n",
       "      <td>16.784959</td>\n",
       "      <td>11.900781</td>\n",
       "      <td>10.495666</td>\n",
       "      <td>10.981432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290812</th>\n",
       "      <td>14.643520</td>\n",
       "      <td>7.362081</td>\n",
       "      <td>7.337679</td>\n",
       "      <td>10.544480</td>\n",
       "      <td>13.553340</td>\n",
       "      <td>10.062991</td>\n",
       "      <td>7.338344</td>\n",
       "      <td>7.064531</td>\n",
       "      <td>7.682399</td>\n",
       "      <td>7.244023</td>\n",
       "      <td>...</td>\n",
       "      <td>8.784437</td>\n",
       "      <td>8.211362</td>\n",
       "      <td>8.560514</td>\n",
       "      <td>12.003981</td>\n",
       "      <td>7.261946</td>\n",
       "      <td>10.468714</td>\n",
       "      <td>17.098442</td>\n",
       "      <td>11.716675</td>\n",
       "      <td>10.132095</td>\n",
       "      <td>9.648341</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 32878 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID_REF      100002    100003     100027     100036     100037     100039    \\\n",
       "GSM290792  13.783817  7.894258   7.836896  11.556448  13.296613   9.808581   \n",
       "GSM290793  14.494117  8.974201   8.454090   9.939125  13.804273  10.497013   \n",
       "GSM290794  13.665953  7.501625   7.278090   9.881257  13.383527   9.261186   \n",
       "GSM290795  14.300739  7.200073  10.258611  10.711247  13.520033  10.150032   \n",
       "GSM290796  13.983886  7.633821   7.134759  10.251259  13.151941   8.837736   \n",
       "GSM290797  14.278781  7.597492   6.972344  11.385409  13.952971   9.372846   \n",
       "GSM290798  14.484616  7.544880   8.021632  11.025483  14.027703  10.217827   \n",
       "GSM290799  14.269023  7.886140   8.680602   9.983204  13.500088   9.915698   \n",
       "GSM290800  13.958387  7.377937   8.281709  11.062754  13.511526   9.133993   \n",
       "GSM290807  14.731631  7.292684   7.531574   9.735370  13.945415  10.044735   \n",
       "GSM290812  14.643520  7.362081   7.337679  10.544480  13.553340  10.062991   \n",
       "\n",
       "ID_REF     100044    100045    100051    100052    ...   10713472  10713880  \\\n",
       "GSM290792  8.411719  7.831912  7.760311  8.393730  ...   9.716142  9.561976   \n",
       "GSM290793  7.647509  8.688864  7.854128  7.523967  ...   8.596880  8.743873   \n",
       "GSM290794  8.182199  8.105078  7.707222  8.101607  ...   8.478933  7.813959   \n",
       "GSM290795  9.329694  7.793425  7.616146  6.122135  ...  10.312558  7.819357   \n",
       "GSM290796  7.437105  7.650994  8.994013  8.454090  ...   8.848244  8.391254   \n",
       "GSM290797  7.572322  8.198980  5.796469  8.626830  ...  10.169555  8.859554   \n",
       "GSM290798  8.433729  8.415782  7.821662  8.449017  ...   8.715933  7.808374   \n",
       "GSM290799  7.422121  8.232527  6.105773  8.972284  ...   9.650487  9.119511   \n",
       "GSM290800  8.156851  9.386527  7.196282  7.830803  ...  10.213736  8.658698   \n",
       "GSM290807  7.083466  8.140264  7.210502  7.964978  ...  10.321017  8.428890   \n",
       "GSM290812  7.338344  7.064531  7.682399  7.244023  ...   8.784437  8.211362   \n",
       "\n",
       "ID_REF      10713947   10714494  10715276   10715317   10715552   10715690  \\\n",
       "GSM290792   8.847066  12.083454  7.325629  10.257842  16.975637  11.864794   \n",
       "GSM290793   7.884095  12.179391  7.986777   9.755269  16.630226  11.887620   \n",
       "GSM290794   8.957044  11.905264  7.278730  10.143787  16.378968  12.334644   \n",
       "GSM290795   8.585799  12.238435  7.469507  10.289238  17.075447  11.750553   \n",
       "GSM290796   7.881447  12.438635  8.713046  10.296148  16.868630  11.692572   \n",
       "GSM290797  10.351108  11.777696  8.004270   9.927231  16.602697  11.770852   \n",
       "GSM290798   7.967301  11.454031  8.248335  10.926243  16.383979  11.793121   \n",
       "GSM290799   8.938449  11.914243  7.232726  10.346964  17.047605  11.785119   \n",
       "GSM290800   9.445913  12.148274  8.022025   9.926582  16.862602  11.528524   \n",
       "GSM290807   7.591398  12.150382  7.318477  10.152613  16.784959  11.900781   \n",
       "GSM290812   8.560514  12.003981  7.261946  10.468714  17.098442  11.716675   \n",
       "\n",
       "ID_REF      10715781   10715908  \n",
       "GSM290792  10.520180  10.913469  \n",
       "GSM290793   9.655440   9.080849  \n",
       "GSM290794  10.128955  11.447662  \n",
       "GSM290795   9.621316  10.624943  \n",
       "GSM290796  10.847043   9.937312  \n",
       "GSM290797  10.529190   8.629565  \n",
       "GSM290798  10.488346  11.264579  \n",
       "GSM290799  10.357344   9.754022  \n",
       "GSM290800  10.634424  11.328658  \n",
       "GSM290807  10.495666  10.981432  \n",
       "GSM290812  10.132095   9.648341  \n",
       "\n",
       "[11 rows x 32878 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ill_1 = df.iloc[29:38,:] \n",
    "df_ill_2 = df.iloc[[44,49],:]\n",
    "df_ill = pd.concat((df_ill_1, df_ill_2))\n",
    "df_ill  #Dataset with just breast cancer people"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7fb34c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tot = pd.concat((df_health, df_ill))  #Complete dataset: healthy + breast cancer people"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aae5ae10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tot.isna().sum().sum()  #We look for NaN in the dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bb3c4ac",
   "metadata": {},
   "source": [
    "#### There are no NaN in our dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86bf4762",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 20 entries, GSM290816 to GSM290812\n",
      "Columns: 32878 entries, 100002 to 10715908\n",
      "dtypes: float64(32878)\n",
      "memory usage: 5.0+ MB\n"
     ]
    }
   ],
   "source": [
    "df_tot.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d4f75862",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID\n",
       "156427      NaN\n",
       "139282    GAP43\n",
       "131316     GPHN\n",
       "235751      NaN\n",
       "696770      NaN\n",
       "          ...  \n",
       "711429      NaN\n",
       "711856      NaN\n",
       "712856      NaN\n",
       "714309      NaN\n",
       "716393      NaN\n",
       "Name: Gene symbol, Length: 32878, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We import now a series which links gene names with they correspondent symbol, which will help us \n",
    "#in subsequent analysis\n",
    "\n",
    "names = pd.read_csv('dataset_gene_name.txt', sep='\\t', index_col='ID', low_memory=False).loc[:,'Gene symbol']\n",
    "names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4f6442ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.series.Series'>\n",
      "Int64Index: 32878 entries, 156427 to 716393\n",
      "Series name: Gene symbol\n",
      "Non-Null Count  Dtype \n",
      "--------------  ----- \n",
      "17916 non-null  object\n",
      "dtypes: object(1)\n",
      "memory usage: 513.7+ KB\n"
     ]
    }
   ],
   "source": [
    "names.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b5d838ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ZKSCAN8</th>\n",
       "      <th>ZNF532</th>\n",
       "      <th>UBE2I</th>\n",
       "      <th>ANAPC7</th>\n",
       "      <th>CAMK1G</th>\n",
       "      <th>EID3</th>\n",
       "      <th>EMC6</th>\n",
       "      <th>ZMYND12</th>\n",
       "      <th>APOBEC3F</th>\n",
       "      <th>EREG</th>\n",
       "      <th>...</th>\n",
       "      <th>RP1L1</th>\n",
       "      <th>ZNF678</th>\n",
       "      <th>SCHIP1</th>\n",
       "      <th>OR2T1</th>\n",
       "      <th>SH2D4B</th>\n",
       "      <th>LILRB2</th>\n",
       "      <th>FAM170A</th>\n",
       "      <th>CUX2</th>\n",
       "      <th>LIN9</th>\n",
       "      <th>MAP6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM290816</th>\n",
       "      <td>8.711001</td>\n",
       "      <td>11.998301</td>\n",
       "      <td>12.888501</td>\n",
       "      <td>8.875158</td>\n",
       "      <td>8.083784</td>\n",
       "      <td>9.526388</td>\n",
       "      <td>13.524342</td>\n",
       "      <td>7.816948</td>\n",
       "      <td>9.270313</td>\n",
       "      <td>8.178023</td>\n",
       "      <td>...</td>\n",
       "      <td>9.476331</td>\n",
       "      <td>10.594055</td>\n",
       "      <td>7.819889</td>\n",
       "      <td>7.079614</td>\n",
       "      <td>7.686465</td>\n",
       "      <td>18.273273</td>\n",
       "      <td>8.075330</td>\n",
       "      <td>8.785462</td>\n",
       "      <td>7.615529</td>\n",
       "      <td>10.737178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290817</th>\n",
       "      <td>7.370765</td>\n",
       "      <td>10.746211</td>\n",
       "      <td>13.683590</td>\n",
       "      <td>10.629111</td>\n",
       "      <td>8.047821</td>\n",
       "      <td>9.767217</td>\n",
       "      <td>13.734776</td>\n",
       "      <td>8.835665</td>\n",
       "      <td>11.061831</td>\n",
       "      <td>8.031646</td>\n",
       "      <td>...</td>\n",
       "      <td>9.203081</td>\n",
       "      <td>9.550889</td>\n",
       "      <td>8.393537</td>\n",
       "      <td>8.772018</td>\n",
       "      <td>8.044503</td>\n",
       "      <td>17.291196</td>\n",
       "      <td>7.669113</td>\n",
       "      <td>7.152983</td>\n",
       "      <td>9.332863</td>\n",
       "      <td>9.639253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290818</th>\n",
       "      <td>7.616559</td>\n",
       "      <td>11.958278</td>\n",
       "      <td>12.248943</td>\n",
       "      <td>7.992173</td>\n",
       "      <td>7.399937</td>\n",
       "      <td>9.875525</td>\n",
       "      <td>13.473040</td>\n",
       "      <td>8.325490</td>\n",
       "      <td>9.554910</td>\n",
       "      <td>7.872808</td>\n",
       "      <td>...</td>\n",
       "      <td>8.219204</td>\n",
       "      <td>11.007826</td>\n",
       "      <td>7.302278</td>\n",
       "      <td>9.167173</td>\n",
       "      <td>8.558747</td>\n",
       "      <td>18.316886</td>\n",
       "      <td>7.481784</td>\n",
       "      <td>9.335869</td>\n",
       "      <td>9.288055</td>\n",
       "      <td>10.567470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290819</th>\n",
       "      <td>7.724920</td>\n",
       "      <td>9.851807</td>\n",
       "      <td>14.121267</td>\n",
       "      <td>9.555641</td>\n",
       "      <td>7.812956</td>\n",
       "      <td>10.055975</td>\n",
       "      <td>13.521172</td>\n",
       "      <td>7.852051</td>\n",
       "      <td>9.804600</td>\n",
       "      <td>7.580368</td>\n",
       "      <td>...</td>\n",
       "      <td>7.706708</td>\n",
       "      <td>10.136614</td>\n",
       "      <td>7.871405</td>\n",
       "      <td>7.788657</td>\n",
       "      <td>7.608933</td>\n",
       "      <td>17.940482</td>\n",
       "      <td>7.675535</td>\n",
       "      <td>10.409149</td>\n",
       "      <td>9.102335</td>\n",
       "      <td>9.191861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290820</th>\n",
       "      <td>7.695143</td>\n",
       "      <td>10.665347</td>\n",
       "      <td>13.814970</td>\n",
       "      <td>9.456169</td>\n",
       "      <td>8.942591</td>\n",
       "      <td>10.196694</td>\n",
       "      <td>13.256555</td>\n",
       "      <td>8.265839</td>\n",
       "      <td>9.913469</td>\n",
       "      <td>7.951397</td>\n",
       "      <td>...</td>\n",
       "      <td>7.793898</td>\n",
       "      <td>10.148575</td>\n",
       "      <td>6.087737</td>\n",
       "      <td>7.628597</td>\n",
       "      <td>8.781164</td>\n",
       "      <td>17.966733</td>\n",
       "      <td>6.814486</td>\n",
       "      <td>9.216344</td>\n",
       "      <td>8.894234</td>\n",
       "      <td>9.433552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290821</th>\n",
       "      <td>8.801931</td>\n",
       "      <td>10.104682</td>\n",
       "      <td>13.650664</td>\n",
       "      <td>10.642495</td>\n",
       "      <td>7.174347</td>\n",
       "      <td>8.443850</td>\n",
       "      <td>13.360218</td>\n",
       "      <td>8.249108</td>\n",
       "      <td>8.759114</td>\n",
       "      <td>8.157565</td>\n",
       "      <td>...</td>\n",
       "      <td>7.210114</td>\n",
       "      <td>8.953207</td>\n",
       "      <td>8.751703</td>\n",
       "      <td>9.764340</td>\n",
       "      <td>8.704036</td>\n",
       "      <td>18.716765</td>\n",
       "      <td>6.923515</td>\n",
       "      <td>9.058786</td>\n",
       "      <td>7.865130</td>\n",
       "      <td>9.838539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290822</th>\n",
       "      <td>9.520554</td>\n",
       "      <td>9.866795</td>\n",
       "      <td>13.529753</td>\n",
       "      <td>9.769843</td>\n",
       "      <td>9.204245</td>\n",
       "      <td>9.119734</td>\n",
       "      <td>13.631693</td>\n",
       "      <td>9.086685</td>\n",
       "      <td>9.604065</td>\n",
       "      <td>7.946977</td>\n",
       "      <td>...</td>\n",
       "      <td>7.313767</td>\n",
       "      <td>11.490777</td>\n",
       "      <td>7.175653</td>\n",
       "      <td>8.437295</td>\n",
       "      <td>8.517987</td>\n",
       "      <td>17.988050</td>\n",
       "      <td>9.520988</td>\n",
       "      <td>9.673355</td>\n",
       "      <td>9.168202</td>\n",
       "      <td>9.778815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290823</th>\n",
       "      <td>7.756522</td>\n",
       "      <td>9.688699</td>\n",
       "      <td>14.372324</td>\n",
       "      <td>10.985775</td>\n",
       "      <td>7.876163</td>\n",
       "      <td>9.606982</td>\n",
       "      <td>13.314560</td>\n",
       "      <td>8.079909</td>\n",
       "      <td>10.569335</td>\n",
       "      <td>7.643912</td>\n",
       "      <td>...</td>\n",
       "      <td>8.588011</td>\n",
       "      <td>9.359134</td>\n",
       "      <td>7.824058</td>\n",
       "      <td>7.820237</td>\n",
       "      <td>8.148928</td>\n",
       "      <td>18.273273</td>\n",
       "      <td>7.388638</td>\n",
       "      <td>8.866602</td>\n",
       "      <td>8.882471</td>\n",
       "      <td>8.290506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290824</th>\n",
       "      <td>8.652816</td>\n",
       "      <td>10.023413</td>\n",
       "      <td>14.179916</td>\n",
       "      <td>10.127081</td>\n",
       "      <td>7.241005</td>\n",
       "      <td>9.577288</td>\n",
       "      <td>13.731312</td>\n",
       "      <td>7.512735</td>\n",
       "      <td>10.023994</td>\n",
       "      <td>7.339665</td>\n",
       "      <td>...</td>\n",
       "      <td>8.417735</td>\n",
       "      <td>9.828095</td>\n",
       "      <td>7.787604</td>\n",
       "      <td>7.107691</td>\n",
       "      <td>8.091075</td>\n",
       "      <td>17.609236</td>\n",
       "      <td>7.759362</td>\n",
       "      <td>8.881503</td>\n",
       "      <td>8.503289</td>\n",
       "      <td>9.700231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290792</th>\n",
       "      <td>7.836896</td>\n",
       "      <td>11.556448</td>\n",
       "      <td>13.296613</td>\n",
       "      <td>9.808581</td>\n",
       "      <td>8.393730</td>\n",
       "      <td>9.142067</td>\n",
       "      <td>12.965830</td>\n",
       "      <td>8.936077</td>\n",
       "      <td>10.720192</td>\n",
       "      <td>9.983204</td>\n",
       "      <td>...</td>\n",
       "      <td>7.426869</td>\n",
       "      <td>9.758695</td>\n",
       "      <td>7.745713</td>\n",
       "      <td>8.150235</td>\n",
       "      <td>8.165984</td>\n",
       "      <td>18.771669</td>\n",
       "      <td>7.331735</td>\n",
       "      <td>9.821140</td>\n",
       "      <td>9.561976</td>\n",
       "      <td>10.520180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290793</th>\n",
       "      <td>8.454090</td>\n",
       "      <td>9.939125</td>\n",
       "      <td>13.804273</td>\n",
       "      <td>10.497013</td>\n",
       "      <td>7.523967</td>\n",
       "      <td>9.362625</td>\n",
       "      <td>13.239098</td>\n",
       "      <td>9.572343</td>\n",
       "      <td>10.886928</td>\n",
       "      <td>8.181475</td>\n",
       "      <td>...</td>\n",
       "      <td>7.885363</td>\n",
       "      <td>8.961994</td>\n",
       "      <td>8.769891</td>\n",
       "      <td>8.552343</td>\n",
       "      <td>7.420828</td>\n",
       "      <td>18.545509</td>\n",
       "      <td>7.663930</td>\n",
       "      <td>7.601909</td>\n",
       "      <td>8.743873</td>\n",
       "      <td>9.655440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290794</th>\n",
       "      <td>7.278090</td>\n",
       "      <td>9.881257</td>\n",
       "      <td>13.383527</td>\n",
       "      <td>9.261186</td>\n",
       "      <td>8.101607</td>\n",
       "      <td>8.923248</td>\n",
       "      <td>13.074373</td>\n",
       "      <td>7.928118</td>\n",
       "      <td>9.894150</td>\n",
       "      <td>7.895988</td>\n",
       "      <td>...</td>\n",
       "      <td>8.051473</td>\n",
       "      <td>9.812328</td>\n",
       "      <td>8.709778</td>\n",
       "      <td>8.041324</td>\n",
       "      <td>7.699717</td>\n",
       "      <td>19.076270</td>\n",
       "      <td>8.174980</td>\n",
       "      <td>9.359250</td>\n",
       "      <td>7.813959</td>\n",
       "      <td>10.128955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290795</th>\n",
       "      <td>10.258611</td>\n",
       "      <td>10.711247</td>\n",
       "      <td>13.520033</td>\n",
       "      <td>10.150032</td>\n",
       "      <td>6.122135</td>\n",
       "      <td>9.998456</td>\n",
       "      <td>13.016399</td>\n",
       "      <td>7.187788</td>\n",
       "      <td>10.841216</td>\n",
       "      <td>8.695254</td>\n",
       "      <td>...</td>\n",
       "      <td>8.666089</td>\n",
       "      <td>11.738657</td>\n",
       "      <td>8.396363</td>\n",
       "      <td>8.904460</td>\n",
       "      <td>7.766847</td>\n",
       "      <td>18.194124</td>\n",
       "      <td>8.911766</td>\n",
       "      <td>9.294017</td>\n",
       "      <td>7.819357</td>\n",
       "      <td>9.621316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290796</th>\n",
       "      <td>7.134759</td>\n",
       "      <td>10.251259</td>\n",
       "      <td>13.151941</td>\n",
       "      <td>8.837736</td>\n",
       "      <td>8.454090</td>\n",
       "      <td>8.966248</td>\n",
       "      <td>13.766642</td>\n",
       "      <td>8.211519</td>\n",
       "      <td>10.357897</td>\n",
       "      <td>8.672921</td>\n",
       "      <td>...</td>\n",
       "      <td>7.790092</td>\n",
       "      <td>11.562495</td>\n",
       "      <td>7.367434</td>\n",
       "      <td>8.760854</td>\n",
       "      <td>7.733277</td>\n",
       "      <td>17.982601</td>\n",
       "      <td>10.218124</td>\n",
       "      <td>9.800906</td>\n",
       "      <td>8.391254</td>\n",
       "      <td>10.847043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290797</th>\n",
       "      <td>6.972344</td>\n",
       "      <td>11.385409</td>\n",
       "      <td>13.952971</td>\n",
       "      <td>9.372846</td>\n",
       "      <td>8.626830</td>\n",
       "      <td>9.633524</td>\n",
       "      <td>13.044780</td>\n",
       "      <td>8.346950</td>\n",
       "      <td>8.543487</td>\n",
       "      <td>8.645184</td>\n",
       "      <td>...</td>\n",
       "      <td>7.947785</td>\n",
       "      <td>10.729930</td>\n",
       "      <td>8.419690</td>\n",
       "      <td>7.562210</td>\n",
       "      <td>8.829405</td>\n",
       "      <td>17.888218</td>\n",
       "      <td>8.003113</td>\n",
       "      <td>9.642482</td>\n",
       "      <td>8.859554</td>\n",
       "      <td>10.529190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290798</th>\n",
       "      <td>8.021632</td>\n",
       "      <td>11.025483</td>\n",
       "      <td>14.027703</td>\n",
       "      <td>10.217827</td>\n",
       "      <td>8.449017</td>\n",
       "      <td>8.911766</td>\n",
       "      <td>13.172426</td>\n",
       "      <td>7.812677</td>\n",
       "      <td>8.696229</td>\n",
       "      <td>8.896148</td>\n",
       "      <td>...</td>\n",
       "      <td>8.743675</td>\n",
       "      <td>10.147113</td>\n",
       "      <td>8.382042</td>\n",
       "      <td>8.969826</td>\n",
       "      <td>8.226455</td>\n",
       "      <td>18.100123</td>\n",
       "      <td>7.694913</td>\n",
       "      <td>8.780831</td>\n",
       "      <td>7.808374</td>\n",
       "      <td>10.488346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290799</th>\n",
       "      <td>8.680602</td>\n",
       "      <td>9.983204</td>\n",
       "      <td>13.500088</td>\n",
       "      <td>9.915698</td>\n",
       "      <td>8.972284</td>\n",
       "      <td>9.030126</td>\n",
       "      <td>13.030929</td>\n",
       "      <td>8.346127</td>\n",
       "      <td>10.423708</td>\n",
       "      <td>9.205629</td>\n",
       "      <td>...</td>\n",
       "      <td>8.449144</td>\n",
       "      <td>10.461621</td>\n",
       "      <td>8.427326</td>\n",
       "      <td>8.588333</td>\n",
       "      <td>7.615233</td>\n",
       "      <td>17.923042</td>\n",
       "      <td>7.887946</td>\n",
       "      <td>9.362625</td>\n",
       "      <td>9.119511</td>\n",
       "      <td>10.357344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290800</th>\n",
       "      <td>8.281709</td>\n",
       "      <td>11.062754</td>\n",
       "      <td>13.511526</td>\n",
       "      <td>9.133993</td>\n",
       "      <td>7.830803</td>\n",
       "      <td>9.405434</td>\n",
       "      <td>12.924767</td>\n",
       "      <td>8.198483</td>\n",
       "      <td>10.134095</td>\n",
       "      <td>9.619962</td>\n",
       "      <td>...</td>\n",
       "      <td>7.823325</td>\n",
       "      <td>9.908717</td>\n",
       "      <td>7.469684</td>\n",
       "      <td>9.132324</td>\n",
       "      <td>9.103808</td>\n",
       "      <td>18.325386</td>\n",
       "      <td>7.419842</td>\n",
       "      <td>10.236709</td>\n",
       "      <td>8.658698</td>\n",
       "      <td>10.634424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290807</th>\n",
       "      <td>7.531574</td>\n",
       "      <td>9.735370</td>\n",
       "      <td>13.945415</td>\n",
       "      <td>10.044735</td>\n",
       "      <td>7.964978</td>\n",
       "      <td>9.830841</td>\n",
       "      <td>12.879289</td>\n",
       "      <td>8.155777</td>\n",
       "      <td>10.018769</td>\n",
       "      <td>9.216510</td>\n",
       "      <td>...</td>\n",
       "      <td>8.270849</td>\n",
       "      <td>10.547232</td>\n",
       "      <td>8.374836</td>\n",
       "      <td>9.283694</td>\n",
       "      <td>7.505329</td>\n",
       "      <td>18.756286</td>\n",
       "      <td>8.211897</td>\n",
       "      <td>10.056449</td>\n",
       "      <td>8.428890</td>\n",
       "      <td>10.495666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290812</th>\n",
       "      <td>7.337679</td>\n",
       "      <td>10.544480</td>\n",
       "      <td>13.553340</td>\n",
       "      <td>10.062991</td>\n",
       "      <td>7.244023</td>\n",
       "      <td>8.996607</td>\n",
       "      <td>13.202428</td>\n",
       "      <td>8.933978</td>\n",
       "      <td>9.240168</td>\n",
       "      <td>9.249096</td>\n",
       "      <td>...</td>\n",
       "      <td>7.044738</td>\n",
       "      <td>11.593240</td>\n",
       "      <td>7.879427</td>\n",
       "      <td>7.357871</td>\n",
       "      <td>8.088103</td>\n",
       "      <td>17.960436</td>\n",
       "      <td>8.109462</td>\n",
       "      <td>10.522151</td>\n",
       "      <td>8.211362</td>\n",
       "      <td>10.132095</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 17916 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ZKSCAN8     ZNF532      UBE2I     ANAPC7    CAMK1G       EID3  \\\n",
       "GSM290816   8.711001  11.998301  12.888501   8.875158  8.083784   9.526388   \n",
       "GSM290817   7.370765  10.746211  13.683590  10.629111  8.047821   9.767217   \n",
       "GSM290818   7.616559  11.958278  12.248943   7.992173  7.399937   9.875525   \n",
       "GSM290819   7.724920   9.851807  14.121267   9.555641  7.812956  10.055975   \n",
       "GSM290820   7.695143  10.665347  13.814970   9.456169  8.942591  10.196694   \n",
       "GSM290821   8.801931  10.104682  13.650664  10.642495  7.174347   8.443850   \n",
       "GSM290822   9.520554   9.866795  13.529753   9.769843  9.204245   9.119734   \n",
       "GSM290823   7.756522   9.688699  14.372324  10.985775  7.876163   9.606982   \n",
       "GSM290824   8.652816  10.023413  14.179916  10.127081  7.241005   9.577288   \n",
       "GSM290792   7.836896  11.556448  13.296613   9.808581  8.393730   9.142067   \n",
       "GSM290793   8.454090   9.939125  13.804273  10.497013  7.523967   9.362625   \n",
       "GSM290794   7.278090   9.881257  13.383527   9.261186  8.101607   8.923248   \n",
       "GSM290795  10.258611  10.711247  13.520033  10.150032  6.122135   9.998456   \n",
       "GSM290796   7.134759  10.251259  13.151941   8.837736  8.454090   8.966248   \n",
       "GSM290797   6.972344  11.385409  13.952971   9.372846  8.626830   9.633524   \n",
       "GSM290798   8.021632  11.025483  14.027703  10.217827  8.449017   8.911766   \n",
       "GSM290799   8.680602   9.983204  13.500088   9.915698  8.972284   9.030126   \n",
       "GSM290800   8.281709  11.062754  13.511526   9.133993  7.830803   9.405434   \n",
       "GSM290807   7.531574   9.735370  13.945415  10.044735  7.964978   9.830841   \n",
       "GSM290812   7.337679  10.544480  13.553340  10.062991  7.244023   8.996607   \n",
       "\n",
       "                EMC6   ZMYND12   APOBEC3F      EREG  ...     RP1L1     ZNF678  \\\n",
       "GSM290816  13.524342  7.816948   9.270313  8.178023  ...  9.476331  10.594055   \n",
       "GSM290817  13.734776  8.835665  11.061831  8.031646  ...  9.203081   9.550889   \n",
       "GSM290818  13.473040  8.325490   9.554910  7.872808  ...  8.219204  11.007826   \n",
       "GSM290819  13.521172  7.852051   9.804600  7.580368  ...  7.706708  10.136614   \n",
       "GSM290820  13.256555  8.265839   9.913469  7.951397  ...  7.793898  10.148575   \n",
       "GSM290821  13.360218  8.249108   8.759114  8.157565  ...  7.210114   8.953207   \n",
       "GSM290822  13.631693  9.086685   9.604065  7.946977  ...  7.313767  11.490777   \n",
       "GSM290823  13.314560  8.079909  10.569335  7.643912  ...  8.588011   9.359134   \n",
       "GSM290824  13.731312  7.512735  10.023994  7.339665  ...  8.417735   9.828095   \n",
       "GSM290792  12.965830  8.936077  10.720192  9.983204  ...  7.426869   9.758695   \n",
       "GSM290793  13.239098  9.572343  10.886928  8.181475  ...  7.885363   8.961994   \n",
       "GSM290794  13.074373  7.928118   9.894150  7.895988  ...  8.051473   9.812328   \n",
       "GSM290795  13.016399  7.187788  10.841216  8.695254  ...  8.666089  11.738657   \n",
       "GSM290796  13.766642  8.211519  10.357897  8.672921  ...  7.790092  11.562495   \n",
       "GSM290797  13.044780  8.346950   8.543487  8.645184  ...  7.947785  10.729930   \n",
       "GSM290798  13.172426  7.812677   8.696229  8.896148  ...  8.743675  10.147113   \n",
       "GSM290799  13.030929  8.346127  10.423708  9.205629  ...  8.449144  10.461621   \n",
       "GSM290800  12.924767  8.198483  10.134095  9.619962  ...  7.823325   9.908717   \n",
       "GSM290807  12.879289  8.155777  10.018769  9.216510  ...  8.270849  10.547232   \n",
       "GSM290812  13.202428  8.933978   9.240168  9.249096  ...  7.044738  11.593240   \n",
       "\n",
       "             SCHIP1     OR2T1    SH2D4B     LILRB2    FAM170A       CUX2  \\\n",
       "GSM290816  7.819889  7.079614  7.686465  18.273273   8.075330   8.785462   \n",
       "GSM290817  8.393537  8.772018  8.044503  17.291196   7.669113   7.152983   \n",
       "GSM290818  7.302278  9.167173  8.558747  18.316886   7.481784   9.335869   \n",
       "GSM290819  7.871405  7.788657  7.608933  17.940482   7.675535  10.409149   \n",
       "GSM290820  6.087737  7.628597  8.781164  17.966733   6.814486   9.216344   \n",
       "GSM290821  8.751703  9.764340  8.704036  18.716765   6.923515   9.058786   \n",
       "GSM290822  7.175653  8.437295  8.517987  17.988050   9.520988   9.673355   \n",
       "GSM290823  7.824058  7.820237  8.148928  18.273273   7.388638   8.866602   \n",
       "GSM290824  7.787604  7.107691  8.091075  17.609236   7.759362   8.881503   \n",
       "GSM290792  7.745713  8.150235  8.165984  18.771669   7.331735   9.821140   \n",
       "GSM290793  8.769891  8.552343  7.420828  18.545509   7.663930   7.601909   \n",
       "GSM290794  8.709778  8.041324  7.699717  19.076270   8.174980   9.359250   \n",
       "GSM290795  8.396363  8.904460  7.766847  18.194124   8.911766   9.294017   \n",
       "GSM290796  7.367434  8.760854  7.733277  17.982601  10.218124   9.800906   \n",
       "GSM290797  8.419690  7.562210  8.829405  17.888218   8.003113   9.642482   \n",
       "GSM290798  8.382042  8.969826  8.226455  18.100123   7.694913   8.780831   \n",
       "GSM290799  8.427326  8.588333  7.615233  17.923042   7.887946   9.362625   \n",
       "GSM290800  7.469684  9.132324  9.103808  18.325386   7.419842  10.236709   \n",
       "GSM290807  8.374836  9.283694  7.505329  18.756286   8.211897  10.056449   \n",
       "GSM290812  7.879427  7.357871  8.088103  17.960436   8.109462  10.522151   \n",
       "\n",
       "               LIN9       MAP6  \n",
       "GSM290816  7.615529  10.737178  \n",
       "GSM290817  9.332863   9.639253  \n",
       "GSM290818  9.288055  10.567470  \n",
       "GSM290819  9.102335   9.191861  \n",
       "GSM290820  8.894234   9.433552  \n",
       "GSM290821  7.865130   9.838539  \n",
       "GSM290822  9.168202   9.778815  \n",
       "GSM290823  8.882471   8.290506  \n",
       "GSM290824  8.503289   9.700231  \n",
       "GSM290792  9.561976  10.520180  \n",
       "GSM290793  8.743873   9.655440  \n",
       "GSM290794  7.813959  10.128955  \n",
       "GSM290795  7.819357   9.621316  \n",
       "GSM290796  8.391254  10.847043  \n",
       "GSM290797  8.859554  10.529190  \n",
       "GSM290798  7.808374  10.488346  \n",
       "GSM290799  9.119511  10.357344  \n",
       "GSM290800  8.658698  10.634424  \n",
       "GSM290807  8.428890  10.495666  \n",
       "GSM290812  8.211362  10.132095  \n",
       "\n",
       "[20 rows x 17916 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We now change column names (gene names) in our dataset with the correspondent gene symbol in the series \"names\"\n",
    "#We discard columns of the dataset which do not have a correspondet name in \"names\"\n",
    "\n",
    "column_names = []\n",
    "mask = names.isna()\n",
    "for col in df_tot.columns:\n",
    "    if mask[col]:\n",
    "        df_tot.drop(columns = col, inplace=True)\n",
    "    else:\n",
    "        column_names.append(names[col])\n",
    "\n",
    "df_tot.columns=column_names\n",
    "df_tot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c916fe34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Finally we separate again our dataset in the healthy and ill patients part\n",
    "\n",
    "df_health = df_tot.iloc[:9,:]\n",
    "df_ill = df_tot.iloc[9:,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "973762ff",
   "metadata": {},
   "source": [
    "#### In this section we look for outliers in our data using the z-score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "39c2dae9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GSM290816    0.949112\n",
       "GSM290817    1.040582\n",
       "GSM290818    0.836374\n",
       "GSM290819    0.671621\n",
       "GSM290820    0.836908\n",
       "GSM290821    0.723389\n",
       "GSM290822    0.827138\n",
       "GSM290823    0.829857\n",
       "GSM290824    0.714762\n",
       "GSM290792    0.879439\n",
       "GSM290793    0.796275\n",
       "GSM290794    0.818664\n",
       "GSM290795    0.667208\n",
       "GSM290796    0.852033\n",
       "GSM290797    0.831959\n",
       "GSM290798    0.868633\n",
       "GSM290799    0.706857\n",
       "GSM290800    0.631493\n",
       "GSM290807    0.685736\n",
       "GSM290812    0.730270\n",
       "dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mean of abs(zscore) for every sample\n",
    "\n",
    "zscore(df_tot).abs().mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ee177931",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['LCA5', 'PCDH20', 'FAM219B', 'PCF11', 'INSL4', 'LRRC39', 'LAMA4',\n",
       "       'MUC4', 'LGSN', 'ECM2',\n",
       "       ...\n",
       "       'C3orf35', 'DNAH12', 'DEFB105A', 'CNOT1', 'KRTAP19-4', 'EYS', 'KANK4',\n",
       "       'OR5K1', 'C15orf53', 'TAS2R20'],\n",
       "      dtype='object', length=633)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Find columns with at least one sample with abs(z-score) > 3\n",
    "\n",
    "df_tot.columns[zscore(df_tot)[zscore(df_tot).abs() > 3].sum(axis=0) > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "1b8fdc07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAGwCAYAAABcnuQpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAvaElEQVR4nO3de1xUdf7H8fcAiqCIq5iAoWBokKALulte+BnlJXNNIk1NUnPVWk1Ls1ravHQRqoddt7KtvFReuhDRPrDUrSytzBQvRUKigVhKtqsCCpIy5/dHP+bXrEAwzjBz8PV8POaB55zvOeczdJp58z3fc47FMAxDAAAAJuXl7gIAAADOB2EGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYmo+7C3A1q9Wqw4cPKyAgQBaLxd3lAACABjAMQ+Xl5QoNDZWXV/19L80+zBw+fFhhYWHuLgMAADjg0KFDuvjii+tt0+zDTEBAgKRffhlt27Z1czUAAKAhysrKFBYWZvser0+zDzM1p5batm1LmAEAwGQaMkSEAcAAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUmv0dgAEAaArV1dXasmWLjhw5opCQECUkJMjb29vdZV0Q6JkBAOA8ZWZmKjIyUomJibrpppuUmJioyMhIZWZmuru0CwJhBgCA85CZmanRo0crNjZWW7duVXl5ubZu3arY2FiNHj2aQNMELIZhGO4uwpXKysoUGBio0tJSHjQJAHCq6upqRUZGKjY2VllZWfLy+v8+AqvVqqSkJOXm5qqgoIBTTo3UmO9vemYAAHDQli1bVFRUpPvuu88uyEiSl5eXUlNTVVhYqC1btripwgsDYQYAAAcdOXJEkhQTE1Pr8pr5Ne3gGoQZAAAcFBISIknKzc2tdXnN/Jp2cA3CDAAADkpISFB4eLjS0tJktVrtllmtVqWnpysiIkIJCQluqvDCQJgBAMBB3t7eevzxx5Wdna2kpCS7q5mSkpKUnZ2tJUuWMPjXxbhpHgAA5yE5OVkZGRm666671L9/f9v8iIgIZWRkKDk52Y3VXRi4NBsAACfgDsDO1Zjvb3pmAABwAm9vb1155ZXuLuOCxJgZAABgaoQZAABgaoQZAABgam4NM5s3b9bIkSMVGhoqi8WirKwsu+WZmZkaOnSoOnToIIvFot27d7ulTgAA4LncGmZOnTql3r1767nnnqtz+cCBA/Xoo482cWUAAMAs3Ho10/DhwzV8+PA6l998882SpKKiogZvs6qqSlVVVbbpsrIyh+sDAACer9mNmUlPT1dgYKDtFRYW5u6SAACACzW7MJOamqrS0lLb69ChQ+4uCQAAuFCzu2mer6+vfH193V0GAABoIs2uZwYAAFxYCDMAAMDU3Hqa6eTJk9q/f79turCwULt371b79u3VpUsXHTt2TMXFxTp8+LAk6dtvv5UkBQcHKzg42C01AwAAz+LWnpkdO3YoLi5OcXFxkqS5c+cqLi5OCxYskCT985//VFxcnEaMGCFJGjdunOLi4vTCCy+4rWYAAOBZLIZhGO4uwpUa8whxAADgGRrz/c2YGQAAYGqEGQAAYGqEGQAAYGqEGQAAYGqEGQAAYGqEGQAAYGrN7tlMAC4M1dXV2rJli44cOaKQkBAlJCTI29vb3WUBcAN6ZgCYTmZmpiIjI5WYmKibbrpJiYmJioyMVGZmprtLA+AGhBkAppKZmanRo0crNjZWW7duVXl5ubZu3arY2FiNHj2aQANcgLgDMADTqK6uVmRkpGJjY5WVlSUvr///e8xqtSopKUm5ubkqKCjglBNgctwBGECztGXLFhUVFem+++6zCzKS5OXlpdTUVBUWFmrLli1uqhCAOxBmAJjGkSNHJEkxMTG1Lq+ZX9MOwIWBMAPANEJCQiRJubm5tS6vmV/TDsCFgTADwDQSEhIUHh6utLQ0Wa1Wu2VWq1Xp6emKiIhQQkKCmyoE4A6EGQCm4e3trccff1zZ2dlKSkqyu5opKSlJ2dnZWrJkCYN/gQsMN80DYCrJycnKyMjQXXfdpf79+9vmR0REKCMjQ8nJyW6sDoA7cGk2AFPiDsBA89aY7296ZgCYkre3t6688kp3lwHAAzBmBgAAmBphBgAAmBphBgAAmBpjZtAgDLYEAHgqembwmzIzMxUZGanExETddNNNSkxMVGRkJE8nBgB4BMIM6pWZmanRo0crNjbW7gZlsbGxGj16NIEGAOB23GcGdaqurlZkZKRiY2OVlZVl95Riq9WqpKQk5ebmqqCggFNOAACnasz3Nz0zqNOWLVtUVFSk++67zy7ISJKXl5dSU1NVWFioLVu2uKlCAAAIM6jHkSNHJEkxMTG1Lq+ZX9MOAAB3IMygTiEhIZKk3NzcWpfXzK9pBwCAOxBmUKeEhASFh4crLS1NVqvVbpnValV6eroiIiKUkJDgpgoBACDMoB7e3t56/PHHlZ2draSkJLurmZKSkpSdna0lS5Yw+BcA4FbcNA/1Sk5OVkZGhu666y7179/fNj8iIkIZGRlKTk52Y3UAAHBpNhqIOwADAJpSY76/6ZlBg3h7e+vKK690dxkAAJyDMTMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDUCDMAAMDU3BpmNm/erJEjRyo0NFQWi0VZWVl2yw3D0IIFCxQSEiI/Pz8NHjxYBQUF7ikWAAB4JLeGmVOnTql379567rnnal3+2GOP6ZlnntELL7ygbdu2qXXr1ho2bJhOnz7dxJUCAABP5ePOnQ8fPlzDhw+vdZlhGHrqqad0//33a9SoUZKkV199VZ06dVJWVpbGjRtX63pVVVWqqqqyTZeVlTm/cAAA4DE8dsxMYWGhSkpKNHjwYNu8wMBAXX755dq6dWud66WnpyswMND2CgsLa4pyAQCAm3hsmCkpKZEkderUyW5+p06dbMtqk5qaqtLSUtvr0KFDLq0TAAC4l1tPM7mCr6+vfH193V0GAABoIh7bMxMcHCxJ+vHHH+3m//jjj7ZlAAAAHhtmIiIiFBwcrA8//NA2r6ysTNu2bVO/fv3cWBkAAPAkbj3NdPLkSe3fv982XVhYqN27d6t9+/bq0qWL7rzzTj388MPq3r27IiIiNH/+fIWGhiopKcl9RQMAAI/i1jCzY8cOJSYm2qbnzp0rSZo0aZJWrlype+65R6dOndL06dN14sQJDRw4UOvXr1erVq3cVTIAAPAwFsMwDHcX4UplZWUKDAxUaWmp2rZt6+5yAABAAzTm+9tjx8wAAAA0BGEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYmkNhZtOmTc6uAwAAwCEOhZlrrrlGl1xyiR5++GEdOnTI2TUBAAA0mENh5ocfftDtt9+ujIwMdevWTcOGDdObb76pn3/+2dn1AQAA1MuhMBMUFKQ5c+Zo9+7d2rZtm3r06KEZM2YoNDRUs2fP1p49e5xdJwAAQK3OewBwfHy8UlNTdfvtt+vkyZNavny5+vTpo4SEBH3zzTfOqBEAAKBODoeZM2fOKCMjQ9dee626du2qDRs26Nlnn9WPP/6o/fv3q2vXrhozZowzawUAADiHxTAMo7ErzZo1S2vXrpVhGLr55ps1depUxcTE2LUpKSlRaGiorFar04p1RFlZmQIDA1VaWqq2bdu6tRYAANAwjfn+9nFkB3v37tXf//53JScny9fXt9Y2QUFBXMINAABczqHTTAsXLtSYMWPOCTJnz57V5s2bJUk+Pj4aNGjQ+VcIAABQD4fCTGJioo4dO3bO/NLSUiUmJp53UQAAAA3lUJgxDEMWi+Wc+f/5z3/UunXr8y4KAACgoRo1ZiY5OVmSZLFYNHnyZLvTTNXV1frqq6/Uv39/pxZYXl6u+fPn65133tHRo0cVFxenp59+Wn/4wx+cuh8AAGBOjQozgYGBkn7pmQkICJCfn59tWcuWLXXFFVdo2rRpTi1w6tSpys3N1WuvvabQ0FCtWrVKgwcP1t69e9W5c2en7gsAAJiPQ5dmP/DAA5o3b57LTylVVlYqICBA7777rkaMGGGb36dPHw0fPlwPP/zwb26DS7MBADAfl1+avXDhQocKa6yzZ8+qurparVq1spvv5+enTz/9tNZ1qqqqVFVVZZsuKytzaY0AAMC9Ghxm4uPj9eGHH+p3v/ud4uLiah0AXGPnzp1OKS4gIED9+vXTQw89pOjoaHXq1Elr167V1q1bFRkZWes66enpeuCBB5yyfwAA4PkaHGZGjRplG/A7atSoesOMM7322muaMmWKOnfuLG9vb8XHx2v8+PHKycmptX1qaqrmzp1rmy4rK1NYWFiT1AoAAJqeQ2Nm3OHUqVMqKytTSEiIxo4dq5MnT2rdunW/uR5jZgAAMJ/GfH87dJ+ZqVOn6uOPP3ZkVYe1bt1aISEhOn78uDZs2KBRo0Y16f4BAIBncijM/PTTT7rmmmsUFhamu+++W3v27HF2XTYbNmzQ+vXrVVhYqH/9619KTExUVFSUbrnlFpftEwAAmIdDYebdd9/VkSNHNH/+fG3fvl3x8fHq2bOn0tLSVFRU5NQCS0tLNXPmTEVFRWnixIkaOHCgNmzYoBYtWjh1PwAAwJycMmbm+++/19q1a7V8+XIVFBTo7NmzzqjNKRgzAwCA+bh8zMyvnTlzRjt27NC2bdtUVFSkTp06ne8mAQAAGszhMLNp0yZNmzZNnTp10uTJk9W2bVtlZ2fr+++/d2Z9AAAA9XLoDsCdO3fWsWPHdM011+jFF1/UyJEj7R46CQAA0FQcCjOLFi3SmDFj1K5dOyeXAwCA56ioqFB+fn6D21dWVqqoqEjh4eF2D2NuiKioKPn7+ze2RMjBMOPsJ2MDAOCJ8vPz1adPnybZV05OjuLj45tkX81Ng8NMcnKyVq5cqbZt2yo5ObnetpmZmeddGAAA7hYVFVXn43Nqk5eXp5SUFK1atUrR0dGN3hcc0+AwExgYaHseU2BgoMsKAgDAU/j7+zvUWxIdHU0vSxNqcJhZsWJFrf8GAABwJ4cuza6srFRFRYVt+uDBg3rqqae0ceNGpxUGAADQEA6FmVGjRunVV1+VJJ04cUJ//OMf9fjjj2vUqFFaunSpUwsEAACoj0NhZufOnUpISJAkZWRkKDg4WAcPHtSrr76qZ555xqkFAgAA1MehMFNRUaGAgABJ0saNG5WcnCwvLy9dccUVOnjwoFMLBAAAqI9DYSYyMlJZWVk6dOiQNmzYoKFDh0qSjh49ysMcAQBAk3IozCxYsEDz5s1TeHi4Lr/8cvXr10/SL700cXFxTi0QAACgPg7dAXj06NEaOHCgjhw5ot69e9vmX3311br++uudVhwAAMBvcSjMSFJwcLCCg4MlSWVlZfroo4906aWXcgdDAADQpBw6zXTjjTfq2WeflfTLPWf69u2rG2+8Ub169dLbb7/t1AIBAADq41CY2bx5s+3S7HfeeUeGYejEiRN65pln9PDDDzu1QAAAgPo4FGZKS0vVvn17SdL69et1ww03yN/fXyNGjFBBQYFTCwQAAKiPQ2EmLCxMW7du1alTp7R+/XrbpdnHjx9Xq1atnFogAABAfRwaAHznnXdqwoQJatOmjbp06aIrr7xS0i+nn2JjY51ZHwAAQL0cCjMzZszQH//4Rx06dEhDhgyRl9cvHTzdunVjzAwAAGhSDl+a3bdvX/Xq1UuFhYW65JJL5OPjoxEjRjizNgAAgN/k8LOZ/vznP8vf3189e/ZUcXGxJGnWrFl65JFHnFogAABAfRwKM6mpqdqzZ48+/vhjuwG/gwcP1htvvOG04gAAAH6LQ6eZsrKy9MYbb+iKK66QxWKxze/Zs6cOHDjgtOIAAAB+i0M9Mz/99JMuuuiic+afOnXKLtwAAAC4mkNhpm/fvlq3bp1tuibAvPzyy7YnaAMAADQFh04zpaWlafjw4dq7d6/Onj2rp59+Wnv37tXnn3+uTz75xNk1AgAA1MmhnpmBAwdqz549Onv2rGJjY7Vx40ZddNFF2rp1q/r06ePsGgEAAOrU6J6ZM2fO6NZbb9X8+fP10ksvuaImAACABmt0z0yLFi309ttvu6IWAACARnNozExSUpKysrI0Z84cZ9cDAIBLFRQUqLy83CXbzsvLs/vpKgEBAerevbtL92EmDoWZ7t2768EHH9Rnn32mPn36qHXr1nbLZ8+e7ZTiAABwpoKCAvXo0cPl+0lJSXH5Pvbt20eg+T8OhZlly5apXbt2ysnJUU5Ojt0yi8VCmAEAeKSaHplVq1YpOjra6duvrKxUUVGRwsPD5efn5/TtS7/0+qSkpLisd8mMHAozhYWFzq4DAIAmEx0drfj4eJdse8CAAS7ZLurm0KXZv2YYhgzDcEYtAAAAjeZwmFm2bJliYmLUqlUrtWrVSjExMXr55ZedWRsAAMBvcug004IFC/TEE09o1qxZtscXbN26VXPmzFFxcbEefPBBpxYJ4MJQUVGh/Pz8Brd3dHxCVFSU/P39HSkRgAdyKMwsXbpUL730ksaPH2+bd91116lXr16aNWsWYQaAQ/Lz85vkLuI5OTkuGy8BoOk5FGbOnDmjvn37njO/T58+Onv27HkXBeDCFBUVdc4VkvWpuaqjsVemREVFOVIeAA/lUJi5+eabtXTpUj3xxBN281988UVNmDDBKYUBuPD4+/s71GPiyitTAHg+h8KM9MsA4I0bN+qKK66QJG3btk3FxcWaOHGi5s6da2v334EHAADAmRwKM7m5uba/gg4cOCBJCgoKUlBQkHJzc23tLBaLE0oEAACom0NhZtOmTc6uAwAAwCEO3Wfmp59+qnPZ119/7XAxAAAAjeVQz0xsbKyWLVumESNG2M1fsmSJ5s+fr8rKSqcUB9dp7P08JO7pAQDwTA6Fmblz5+qGG27QLbfcoieeeELHjh3TxIkT9fXXX2vNmjXOrhEu0FT385C4pwcAwLUcCjP33HOPhgwZoptvvlm9evXSsWPHdPnll+urr75ScHCws2uECzT2fh4S9/QAAHgmhy/NjoyMVExMjN5++21J0tixYwkyJuLo/Twk7ukBAPAsDg0A/uyzz9SrVy8VFBToq6++0tKlSzVr1iyNHTtWx48fd3aNAAAAdXIozFx11VUaO3asvvjiC0VHR2vq1KnatWuXiouLFRsb6+waAQAA6uTQaaaNGzdq0KBBdvMuueQSffbZZ1q8eLFTCgMAAGiIRvXMXHvttSotLbUFmUceeUQnTpywLT9+/LjWrl3r1AIBAADq06gws2HDBlVVVdmm09LSdOzYMdv02bNn9e233zqvOgAAgN/QqNNMhmHUO+1s1dXVWrRokVatWqWSkhKFhoZq8uTJuv/++3nuEwCg0SxnTysu2Et+J/ZJhx0aNup2fif2KS7YS5azp91disdw+NLspvDoo49q6dKleuWVV9SzZ0/t2LFDt9xyiwIDAzV79mx3lwcAMJlWJ4u189Y20uZbpc3ursYx0ZJ23tpGeSeLJfV3dzkeoVFhxmKxnNMj4soeks8//1yjRo2yPTYhPDxca9eu1ZdffumyfQIAmq/Tbboo/h8ntXr1akWb9Iaeefn5mjBhgpZd28XdpXiMRp9mmjx5snx9fSVJp0+f1m233abWrVtLkt14Gmfo37+/XnzxRe3bt089evTQnj179Omnn+qJJ56oc52qqiq7OsrKypxaEwDAvAyfVtpVYlVlux5S6O/dXY5DKkus2lVileHTyt2leIxGhZlJkybZTaekpJzTZuLEiedX0a/89a9/VVlZmaKiouTt7a3q6motXrxYEyZMqHOd9PR0PfDAA06rAQAAeLZGhZkVK1a4qo5avfnmm1q9erXWrFmjnj17avfu3brzzjsVGhp6TrCqkZqaqrlz59qmy8rKFBYW1lQlAwCAJubRA4Dvvvtu/fWvf9W4ceMkSbGxsTp48KDS09PrDDO+vr6202AA3KugoEDl5eUu235eXp7dT1cJCAhQ9+7dXboPAI7z6DBTUVEhLy/7S+e8vb1ltVrdVBGAhiooKFCPHj2aZF+1nfJ2tn379hFoAA/l0WFm5MiRWrx4sbp06aKePXtq165deuKJJzRlyhR3lwbgN9T0yKxatUrR0dEu2UdlZaWKiooUHh4uPz8/l+wjLy9PKSkpLu1hAnB+PDrM/P3vf9f8+fM1Y8YMHT16VKGhobr11lu1YMECd5cGoIGio6MVHx/vsu0PGDDAZdsGYA4eHWYCAgL01FNP6amnnnJ3KQAAwEOZ817OAAAA/8eje2YAAHCmiooKSdLOnTtdsv2mGscFe4QZAMAFIz8/X5I0bdo0N1dy/gICAtxdgscgzAAALhhJSUmSpKioKPn7+zt9+zVXv7nyKj6Jex/9N8IMAOCCERQUpKlTp7p8P66+ig/2GAAMAABMjTADAABMjTADAABMjTADAABMjTADAABMjTADAABMjTADAABMjTADAABMjZvmNSMFBQUqLy932fZrngfiyueCcFdLAEBjEWaaiYKCAvXo0aNJ9pWSkuLS7e/bt49AAwBoMMJMM1HTI+PK54G4+mmwNc80cWXvEgCg+SHMNDOufh7IgAEDXLZtAAAcQZgB4BKWs6cVF+wlvxP7pMPmvdbA78Q+xQV7yXL2tLtLAVAHwgwAl2h1slg7b20jbb5V2uzuahwXLWnnrW2Ud7JYUn93lwOgFoQZAC5xuk0Xxf/jpFavXq3oqCh3l+OwvPx8TZgwQcuu7eLuUgDUgTADwCUMn1baVWJVZbseUujv3V2OwypLrNpVYpXh08rdpQCog3lPZAMAAIgwAwAATI4wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATI0wAwAATM3H3QXAOSxnTysu2Et+J/ZJh82ZUf1O7FNcsJcsZ0+7uxQAgIkQZpqJVieLtfPWNtLmW6XN7q7GMdGSdt7aRnkniyX1d3c5AACTIMw0E6fbdFH8P05q9erVio6Kcnc5DsnLz9eECRO07Nou7i4FAGAihJlmwvBppV0lVlW26yGF/t7d5TikssSqXSVWGT6t3F0KAMBEzDm4AgAA4P8QZgAAgKkRZgAAgKkRZgAAgKkRZgAAgKlxNVMzUVFRIUnauXOny/ZRWVmpoqIihYeHy8/Pz+nbz8vLc/o24T7N4ZiUOC4BMyDMNBP5+fmSpGnTprm5kvMXEBDg7hLgBM3pmJQ4LgFPRphpJpKSkiRJUVFR8vf3d8k+8vLylJKSolWrVik6Otol+wgICFD37t1dsm00reZyTEocl4CnI8w0E0FBQZo6dWqT7Cs6Olrx8fFNsi+YF8ckgKbCAGAAAGBq9MwAAFCHiooK2/ivhqgZMO7IwHFXnpJt7ggzAADUIT8/X3369Gn0eikpKY1eJycnh9OlDiLMAABQh6ioKOXk5DS4/fncLiAqKqqx5eH/EGYAAKiDv79/o3tLBgwY4KJqUBcGAAMAAFPz+DATHh4ui8VyzmvmzJnuLg0AAHgAjz/NtH37dlVXV9umc3NzNWTIEI0ZM8aNVQEAAE/h8WGmY8eOdtOPPPKILrnkEg0aNMhNFQEAAE/i8WHm137++WetWrVKc+fOlcViqbVNVVWVqqqqbNNlZWVNVR4AAHADjx8z82tZWVk6ceKEJk+eXGeb9PR0BQYG2l5hYWFNVyAAAGhypgozy5Yt0/DhwxUaGlpnm9TUVJWWltpehw4dasIKAQBAUzPNaaaDBw/qgw8+UGZmZr3tfH195evr20RVAQAAdzNNz8yKFSt00UUXacSIEe4uBQAAeBBThBmr1aoVK1Zo0qRJ8vExTWcSAABoAqYIMx988IGKi4s1ZcoUd5cCAAA8jCm6OYYOHSrDMNxdBgAA8ECm6JkBAACoC2EGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYGmEGAACYmo+7C4B7VFRUKD8/v1Hr5OXl2f1sqKioKPn7+zdqHQAAGoowc4HKz89Xnz59HFo3JSWlUe1zcnIUHx/v0L4AAPgthJkLVFRUlHJychq1TmVlpYqKihQeHi4/P79G7QsAAFchzFyg/P39HeotGTBggAuqAQDAcQwABgAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApkaYAQAApubj7gIAoEZFRYXy8/Mb3D4vL8/uZ0NFRUXJ39+/UesA8FyEGQAeIz8/X3369Gn0eikpKY1qn5OTo/j4+EbvB4BnIswA8BhRUVHKyclpcPvKykoVFRUpPDxcfn5+jdoPgObDYhiG4e4iXKmsrEyBgYEqLS1V27Zt3V0OAABogMZ8f3v8AOAffvhBKSkp6tChg/z8/BQbG6sdO3a4uywAAOAhPPo00/HjxzVgwAAlJibq/fffV8eOHVVQUKDf/e537i4NAAB4CI8OM48++qjCwsK0YsUK27yIiAg3VgQAADyNR59m+uc//6m+fftqzJgxuuiiixQXF6eXXnqp3nWqqqpUVlZm9wIAAM2XR4eZ7777TkuXLlX37t21YcMG/eUvf9Hs2bP1yiuv1LlOenq6AgMDba+wsLAmrBgAADQ1j76aqWXLlurbt68+//xz27zZs2dr+/bt2rp1a63rVFVVqaqqyjZdVlamsLAwrmYCAMBEms3VTCEhIbrsssvs5kVHR6u4uLjOdXx9fdW2bVu7FwAAaL48OswMGDBA3377rd28ffv2qWvXrm6qCAAAeBqPDjNz5szRF198obS0NO3fv19r1qzRiy++qJkzZ7q7NAAA4CE8Osz84Q9/0DvvvKO1a9cqJiZGDz30kJ566ilNmDDB3aUBAAAP4dEDgJ2BxxkAAGA+zWYAMAAAwG8hzAAAAFPz6McZOEPNWTTuBAwAgHnUfG83ZDRMsw8z5eXlksSdgAEAMKHy8nIFBgbW26bZDwC2Wq06fPiwAgICZLFY3F2OqdXcTfnQoUMMpoZH4JiEp+GYdB7DMFReXq7Q0FB5edU/KqbZ98x4eXnp4osvdncZzQp3Voan4ZiEp+GYdI7f6pGpwQBgAABgaoQZAABgaoQZNJivr68WLlwoX19fd5cCSOKYhOfhmHSPZj8AGAAANG/0zAAAAFMjzAAAAFMjzAAAAFMjzAAAAFMjzFxAJk+erKSkpDqX79q1S2PGjFGnTp3UqlUrde/eXdOmTdO+ffvOaTts2DB5e3tr+/bt5yxbtGiRLBaL3SsqKsqZbwUebPLkybb/7i1atFCnTp00ZMgQLV++XFar1dYuPDxcFotFX3zxhd36d955p6688spztvv999+rZcuWiomJqXW/FotFWVlZtdbz6+N+8+bNGjlypEJDQ2td58yZM7r33nsVGxur1q1bKzQ0VBMnTtThw4cb/DuAOf362P3165prrrG1+a3PyaKiIlksFnl7e+uHH36w2/6RI0fk4+Mji8WioqIiSdKePXs0fvx4hYWFyc/PT9HR0Xr66aeb7D03F4QZSJKys7N1xRVXqKqqSqtXr1ZeXp5WrVqlwMBAzZ8/365tcXGxPv/8c91+++1avnx5rdvr2bOnjhw5Ynt9+umnTfE24CGuueYaHTlyREVFRXr//feVmJioO+64Q3/605909uxZW7tWrVrp3nvvbdA2V65cqRtvvFFlZWXatm2bw7WdOnVKvXv31nPPPVfr8oqKCu3cuVPz58/Xzp07lZmZqW+//VbXXXedw/uEedQcu79+rV27VlLjPic7d+6sV1991W7eK6+8os6dO9vNy8nJ0UUXXaRVq1bpm2++0d/+9jelpqbq2Wefde0bbW4MXDAmTZpkjBo16pz5p06dMoKCgoykpKRa1zt+/Ljd9KJFi4xx48YZeXl5RmBgoFFRUWG3fOHChUbv3r2dVDXMpq7j7MMPPzQkGS+99JJhGIbRtWtXY/bs2UbLli2NdevW2drdcccdxqBBg+zWtVqtRrdu3Yz169cb9957rzFt2rRzti/JeOeddxpcT33r/Lcvv/zSkGQcPHjwN9vCvOo7Vhr6OVlYWGhIMu6//36je/fudm169OhhzJ8/35BkFBYW1lnHjBkzjMTEREfewgWLnhlow4YN+ve//6177rmn1uXt2rWz/dswDK1YsUIpKSmKiopSZGSkMjIyzlmnoKBAoaGh6tatmyZMmKDi4mJXlQ+TuOqqq9S7d29lZmba5kVEROi2225Tamqq3Smo/7Zp0yZVVFRo8ODBSklJ0euvv65Tp041RdmSpNLSUlksFrv/F3BhacznpCRdd911On78uK1X+tNPP9Xx48c1cuTI39xXaWmp2rdvf941X0gIM1BBQYEkNWhcywcffKCKigoNGzZMkpSSkqJly5bZtbn88su1cuVKrV+/XkuXLlVhYaESEhJUXl7u/OJhKlFRUbaxAjXuv/9+FRYWavXq1XWut2zZMo0bN07e3t6KiYlRt27d9NZbb53Tbvz48WrTpo3dq77tNsTp06d17733avz48Tw48AKQnZ19zjGUlpbWqM9JSWrRooVSUlJsp+KXL1+ulJQUtWjRot71Pv/8c73xxhuaPn36+b2RCwxhBjIacRPo5cuXa+zYsfLx+eWB6+PHj9dnn32mAwcO2NoMHz5cY8aMUa9evTRs2DC99957OnHihN58802n1w5zMQxDFovFbl7Hjh01b948LViwQD///PM565w4cUKZmZlKSUmxzastREvSk08+qd27d9u9zmesy5kzZ3TjjTfKMAwtXbrU4e3APBITE885hm677bZGfU7WmDJlit566y2VlJTorbfe0pQpU+ptn5ubq1GjRmnhwoUaOnSoo2/hguTj7gLgfj169JAk5efnq1+/fnW2O3bsmN555x2dOXPG7oO9urpay5cv1+LFi2tdr127durRo4f279/v3MJhOnl5eYqIiDhn/ty5c/X888/r+eefP2fZmjVrdPr0aV1++eW2eYZhyGq1at++fbbjV5KCg4MVGRlpt35AQIBOnDjR6FprgszBgwf10Ucf0StzgWjduvU5x5DU8M/JX4uNjVVUVJTGjx+v6OhoxcTEaPfu3bW23bt3r66++mpNnz5d999/v8P1X6jomYGGDh2qoKAgPfbYY7Uur/kiWL16tS6++GLt2bPH7q+Wxx9/XCtXrlR1dXWt6588eVIHDhxQSEiIq94CTOCjjz7S119/rRtuuOGcZW3atNH8+fO1ePHic05HLlu2THfddZfdMbdnzx4lJCTUeTXd+aoJMgUFBfrggw/UoUMHl+wH5tHQz8n/NmXKFH388cf19sp88803SkxM1KRJk+r8oxD1o2fmAlNaWnrOXwYdOnTQyy+/rDFjxui6667T7NmzFRkZqX//+9968803VVxcrNdff13Lli3T6NGjz7nPR1hYmFJTU7V+/XqNGDFC8+bN08iRI9W1a1cdPnxYCxculLe3t8aPH9+E7xTuVFVVpZKSElVXV+vHH3/U+vXrlZ6erj/96U+aOHFiretMnz5dTz75pNasWWPrhdm9e7d27typ1atXnzNWYfz48XrwwQf18MMP2057NsTJkyftegkLCwu1e/dutW/fXl26dNGZM2c0evRo7dy5U9nZ2aqurlZJSYkkqX379mrZsmVjfx0wkZpj99d8fHwUFBTUoM/J/zZt2jSNGTOmzsHjubm5uuqqqzRs2DDNnTvXtm9vb2917NjR6e+v2XLnpVRoWpMmTTIknfP685//bBiGYWzfvt1ITk42OnbsaPj6+hqRkZHG9OnTjYKCAmPHjh2GJOPLL7+sddvDhw83rr/+esMwDGPs2LFGSEiI0bJlS6Nz587G2LFjjf379zfZ+4R7/fo48/HxMTp27GgMHjzYWL58uVFdXW1r17VrV+PJJ5+0W3fNmjWGJNul2bfffrtx2WWX1bqfI0eOGF5eXsa7775rGEbDL83etGlTrf8fTJo0yTCM/7+0trbXpk2bHP21wATq+oy89NJLbW3q+5w0jP8/fnbt2lXrPnbt2mV3afbChQtr3WfXrl1d/G6bF4thODCqCQAAwEMwZgYAAJgaYQYAAJgaYQYAAJgaYQYAAJgaYQYAAJgaYQYAAJgaYQYAAJgaYQYAAJgaYQYAAJgaYQZAkyspKdEdd9yhyMhItWrVSp06ddKAAQO0dOlSVVRUuLs8ACbDgyYBNKnvvvtOAwYMULt27ZSWlqbY2Fj5+vrq66+/1osvvqjOnTvruuuuc3eZAEyEnhkATWrGjBny8fHRjh07dOONNyo6OlrdunXTqFGjtG7dOo0cOVKSdOLECU2dOlUdO3ZU27ZtddVVV2nPnj227SxatEi///3v9dprryk8PFyBgYEaN26cysvLbW2sVqvS09MVEREhPz8/9e7dWxkZGbblx48f14QJE9SxY0f5+fmpe/fuWrFiRdP9MgA4BWEGQJP5z3/+o40bN2rmzJlq3bp1rW0sFoskacyYMTp69Kjef/995eTkKD4+XldffbWOHTtma3vgwAFlZWUpOztb2dnZ+uSTT/TII4/Ylqenp+vVV1/VCy+8oG+++UZz5sxRSkqKPvnkE0nS/PnztXfvXr3//vvKy8vT0qVLFRQU5MLfAABX4DQTgCazf/9+GYahSy+91G5+UFCQTp8+LUmaOXOmRo4cqS+//FJHjx6Vr6+vJGnJkiXKyspSRkaGpk+fLumXnpeVK1cqICBAknTzzTfrww8/1OLFi1VVVaW0tDR98MEH6tevnySpW7du+vTTT/WPf/xDgwYNUnFxseLi4tS3b19JUnh4eFP8GgA4GWEGgNt9+eWXslqtmjBhgqqqqrRnzx6dPHlSHTp0sGtXWVmpAwcO2KbDw8NtQUaSQkJCdPToUUm/BKeKigoNGTLEbhs///yz4uLiJEl/+ctfdMMNN2jnzp0aOnSokpKS1L9/f1e9TQAuQpgB0GQiIyNlsVj07bff2s3v1q2bJMnPz0+SdPLkSYWEhOjjjz8+Zxvt2rWz/btFixZ2yywWi6xWq20bkrRu3Tp17tzZrl1Nb8/w4cN18OBBvffee/rXv/6lq6++WjNnztSSJUscf5MAmhxhBkCT6dChg4YMGaJnn31Ws2bNqnPcTHx8vEpKSuTj4+PwqZ/LLrtMvr6+Ki4u1qBBg+ps17FjR02aNEmTJk1SQkKC7r77bsIMYDKEGQBN6vnnn9eAAQPUt29fLVq0SL169ZKXl5e2b9+u/Px89enTR4MHD1a/fv2UlJSkxx57TD169NDhw4e1bt06XX/99bYxLvUJCAjQvHnzNGfOHFmtVg0cOFClpaX67LPP1LZtW02aNEkLFixQnz591LNnT1VVVSk7O1vR0dFN8FsA4EyEGQBN6pJLLtGuXbuUlpam1NRUff/99/L19dVll12mefPmacaMGbJYLHrvvff0t7/9Tbfccot++uknBQcH63/+53/UqVOnBu/roYceUseOHZWenq7vvvtO7dq1U3x8vO677z5JUsuWLZWamqqioiL5+fkpISFBr7/+uqveOgAXsRiGYbi7CAAAAEdxnxkAAGBqhBkAAGBqhBkAAGBqhBkAAGBqhBkAAGBqhBkAAGBqhBkAAGBqhBkAAGBqhBkAAGBqhBkAAGBqhBkAAGBq/wsS4BrmmLegbgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Boxplot of some genes with outliers\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.boxplot(df_tot[['LCA5','DNAH12', 'ECM2']], labels=['LCA5','DNAH12', 'ECM2'])\n",
    "ax.set_xlabel('Genes')\n",
    "ax.set_ylabel('Expressivity')\n",
    "#plt.show()\n",
    "plt.savefig('boxplot_genes.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcf79f58",
   "metadata": {},
   "source": [
    "## Statistical analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e5904caa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#First of all we set a threshold value for the type I error\n",
    "\n",
    "alpha = 0.05  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f91e5909",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We perform a t-test t find differently expressed genes among healthy and ill people\n",
    "\n",
    "stat, p_values = ttest_ind(df_ill.values, df_health.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f5876a5",
   "metadata": {},
   "source": [
    "#### When stat is positive: tumor patients have over-expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c4942dfa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2828"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Number of differentially expressed genes with statistical significance (we choose to set type I error alpha = 0.05)\n",
    "\n",
    "len(p_values[p_values<alpha])  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e5b4090f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1171, 1657)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We divide differentially expressed genes in overexpressed and underexpressed (according to the t-test)\n",
    "\n",
    "gene_signif = df_tot.columns[p_values<0.05]  #symbols of differentially expressed genes\n",
    "stat_signif = stat[p_values<0.05]  #t-test statistics of differentially expressed genes\n",
    "overexpr_gene = gene_signif[stat_signif>0]  #when t-test statistic is positive: overexpressed genes in ill patients\n",
    "underexpr_gene = gene_signif[stat_signif<0]  #when t-test statistic is negative: underexpressed genes in ill patients\n",
    "len(overexpr_gene),len(underexpr_gene)   #number of over and under expressed genes after p-values correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f922988c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We coorect p-values for multiple testing applying the Hochberg-Benjamini procedure\n",
    "\n",
    "rej, p_adj, _, _ = multipletests(p_values, alpha = alpha, method='fdr_bh')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "49e8b7a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "433"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Number of differentially expressed genes with statistical significance after correcting p-values for multiple \n",
    "#testing (we choose to set type I error alpha = 0.05)\n",
    "\n",
    "len(p_adj[p_adj<alpha])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c5b44487",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(92, 341)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We divide differentially expressed genes in overexpressed and underexpressed (according to the t-test)\n",
    "\n",
    "diff_gen = df_tot.columns[rej]  #symbols of differentially expressed genes\n",
    "diff_stat = stat[p_adj<alpha]  #t-test statistics of differentially expressed genes\n",
    "gene_up = list(diff_gen[diff_stat>0])  #when t-test statistic is positive: overexpressed genes in ill patients\n",
    "gene_down = list(diff_gen[diff_stat<0])  #when t-test statistic is negative: underexpressed genes in ill patients\n",
    "(len(gene_up), len(gene_down))  #number of over and under expressed genes after p-values correction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c194b3d",
   "metadata": {},
   "source": [
    "## Over expression enrichment analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7324da3",
   "metadata": {},
   "source": [
    "### Under-expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4c762453",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1550</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Metabolism Of RNA R-HSA-8953854</td>\n",
       "      <td>55/666</td>\n",
       "      <td>6.971826e-24</td>\n",
       "      <td>5.570489e-21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.452752</td>\n",
       "      <td>344.061811</td>\n",
       "      <td>EIF4A2;POP5;HBS1L;RPL30;RPL31;CWC27;RPL7;RPS17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1551</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Translation R-HSA-72766</td>\n",
       "      <td>36/281</td>\n",
       "      <td>3.933964e-22</td>\n",
       "      <td>1.571619e-19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10.019678</td>\n",
       "      <td>493.842107</td>\n",
       "      <td>EIF4A2;SRP19;RPL30;RPL31;MRPS14;RPL36A;MRPL17;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>BioPlanet_2019</td>\n",
       "      <td>Translation</td>\n",
       "      <td>27/151</td>\n",
       "      <td>1.038130e-20</td>\n",
       "      <td>6.218399e-18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14.482801</td>\n",
       "      <td>666.415680</td>\n",
       "      <td>EIF4A2;SRP19;RPL30;RPL31;RPL36A;SRP14;RPL7;RPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2349</th>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "      <td>cotranslational protein targeting to membrane ...</td>\n",
       "      <td>22/94</td>\n",
       "      <td>8.968832e-20</td>\n",
       "      <td>1.609905e-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20.036882</td>\n",
       "      <td>878.776505</td>\n",
       "      <td>SRP19;RPL30;RPL41;RPL21;RPS7;RPL31;RPL23;RPL36...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2351</th>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "      <td>protein targeting to ER (GO:0045047)</td>\n",
       "      <td>22/103</td>\n",
       "      <td>7.577069e-19</td>\n",
       "      <td>4.533613e-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17.802387</td>\n",
       "      <td>742.786605</td>\n",
       "      <td>SRP19;RPL30;RPL41;RPL21;RPS7;RPL31;RPL23;RPL36...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1681</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Downregulation Of TGF-beta Receptor Signaling ...</td>\n",
       "      <td>3/26</td>\n",
       "      <td>8.101664e-03</td>\n",
       "      <td>4.867090e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.062346</td>\n",
       "      <td>38.825726</td>\n",
       "      <td>PPP1CB;SMAD2;PPP1CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1682</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Insulin Receptor Recycling R-HSA-77387</td>\n",
       "      <td>3/26</td>\n",
       "      <td>8.101664e-03</td>\n",
       "      <td>4.867090e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.062346</td>\n",
       "      <td>38.825726</td>\n",
       "      <td>ATP6V1G1;ATP6V1E1;ATP6V1D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1683</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>FCERI Mediated NF-kB Activation R-HSA-2871837</td>\n",
       "      <td>5/78</td>\n",
       "      <td>8.390046e-03</td>\n",
       "      <td>4.933773e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.249610</td>\n",
       "      <td>20.316149</td>\n",
       "      <td>PSMA3;PSMC6;PSMD14;PSMA2;SKP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1684</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Nuclear Events Mediated By NFE2L2 R-HSA-9759194</td>\n",
       "      <td>5/78</td>\n",
       "      <td>8.390046e-03</td>\n",
       "      <td>4.933773e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.249610</td>\n",
       "      <td>20.316149</td>\n",
       "      <td>PSMA3;PSMC6;PSMD14;PSMA2;SKP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1685</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Regulation Of Ornithine Decarboxylase (ODC) R-...</td>\n",
       "      <td>4/50</td>\n",
       "      <td>8.397912e-03</td>\n",
       "      <td>4.933773e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.385544</td>\n",
       "      <td>25.741673</td>\n",
       "      <td>PSMA3;PSMC6;PSMD14;PSMA2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Gene_set  \\\n",
       "1550               Reactome_2022   \n",
       "1551               Reactome_2022   \n",
       "47                BioPlanet_2019   \n",
       "2349  GO_Biological_Process_2021   \n",
       "2351  GO_Biological_Process_2021   \n",
       "...                          ...   \n",
       "1681               Reactome_2022   \n",
       "1682               Reactome_2022   \n",
       "1683               Reactome_2022   \n",
       "1684               Reactome_2022   \n",
       "1685               Reactome_2022   \n",
       "\n",
       "                                                   Term Overlap       P-value  \\\n",
       "1550                    Metabolism Of RNA R-HSA-8953854  55/666  6.971826e-24   \n",
       "1551                            Translation R-HSA-72766  36/281  3.933964e-22   \n",
       "47                                          Translation  27/151  1.038130e-20   \n",
       "2349  cotranslational protein targeting to membrane ...   22/94  8.968832e-20   \n",
       "2351               protein targeting to ER (GO:0045047)  22/103  7.577069e-19   \n",
       "...                                                 ...     ...           ...   \n",
       "1681  Downregulation Of TGF-beta Receptor Signaling ...    3/26  8.101664e-03   \n",
       "1682             Insulin Receptor Recycling R-HSA-77387    3/26  8.101664e-03   \n",
       "1683      FCERI Mediated NF-kB Activation R-HSA-2871837    5/78  8.390046e-03   \n",
       "1684    Nuclear Events Mediated By NFE2L2 R-HSA-9759194    5/78  8.390046e-03   \n",
       "1685  Regulation Of Ornithine Decarboxylase (ODC) R-...    4/50  8.397912e-03   \n",
       "\n",
       "      Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
       "1550      5.570489e-21            0                     0    6.452752   \n",
       "1551      1.571619e-19            0                     0   10.019678   \n",
       "47        6.218399e-18            0                     0   14.482801   \n",
       "2349      1.609905e-16            0                     0   20.036882   \n",
       "2351      4.533613e-16            0                     0   17.802387   \n",
       "...                ...          ...                   ...         ...   \n",
       "1681      4.867090e-02            0                     0    8.062346   \n",
       "1682      4.867090e-02            0                     0    8.062346   \n",
       "1683      4.933773e-02            0                     0    4.249610   \n",
       "1684      4.933773e-02            0                     0    4.249610   \n",
       "1685      4.933773e-02            0                     0    5.385544   \n",
       "\n",
       "      Combined Score                                              Genes  \n",
       "1550      344.061811  EIF4A2;POP5;HBS1L;RPL30;RPL31;CWC27;RPL7;RPS17...  \n",
       "1551      493.842107  EIF4A2;SRP19;RPL30;RPL31;MRPS14;RPL36A;MRPL17;...  \n",
       "47        666.415680  EIF4A2;SRP19;RPL30;RPL31;RPL36A;SRP14;RPL7;RPS...  \n",
       "2349      878.776505  SRP19;RPL30;RPL41;RPL21;RPS7;RPL31;RPL23;RPL36...  \n",
       "2351      742.786605  SRP19;RPL30;RPL41;RPL21;RPS7;RPL31;RPL23;RPL36...  \n",
       "...              ...                                                ...  \n",
       "1681       38.825726                                PPP1CB;SMAD2;PPP1CC  \n",
       "1682       38.825726                          ATP6V1G1;ATP6V1E1;ATP6V1D  \n",
       "1683       20.316149                      PSMA3;PSMC6;PSMD14;PSMA2;SKP1  \n",
       "1684       20.316149                      PSMA3;PSMC6;PSMD14;PSMA2;SKP1  \n",
       "1685       25.741673                           PSMA3;PSMC6;PSMD14;PSMA2  \n",
       "\n",
       "[259 rows x 10 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We perform over-representation analysis using the python package gseapy\n",
    "#We start with underexpressed genes\n",
    "\n",
    "enr_down = gp.enrich(gene_list=gene_down, \n",
    "                gene_sets=['BioCarta_2016', 'BioPlanet_2019', 'WikiPathway_2021_Human',\\\n",
    "                           'Elsevier_Pathway_Collection', 'KEGG_2021_Human', 'Panther_2016', 'Reactome_2022',\\\n",
    "                           'GO_Biological_Process_2021'], \n",
    "                background=df_tot.columns)\n",
    "\n",
    "#REM: by setting a backgrounf value in the function we are giving the function the complex of genes we have \n",
    "#considered in our statistical analysis. The goal is to obtain more accurate values for the statistic of the \n",
    "#over-representation test performed by the function \"enrich\"\n",
    "\n",
    "#We now look for pathways which have a significant statistical over-representation\n",
    "enr_down.results[enr_down.results['Adjusted P-value']<alpha].sort_values(by=['Adjusted P-value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "bf88a0c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We plot the best results of over-representation analysis on under-expressed genes \n",
    "\n",
    "ax = dotplot(enr_down.results,\n",
    "              column=\"Adjusted P-value\",\n",
    "              x='Gene_set', # set x axis, so you could do a multi-sample/library comparsion\n",
    "              size=10,\n",
    "              top_term=3,\n",
    "              figsize=(3,5),\n",
    "              title = \"UNDER EXPRESSED\",\n",
    "              xticklabels_rot=45, # rotate xtick labels\n",
    "              show_ring=True, # set to False to revmove outer ring\n",
    "              marker='o',\n",
    "              ofname='down_genes_dotlot.png'\n",
    "             )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0ddc754",
   "metadata": {},
   "source": [
    "### Over-expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b3a3fd95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Activation Of Gene Expression By SREBF (SREBP)...</td>\n",
       "      <td>4/42</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.010383</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24.042347</td>\n",
       "      <td>243.877579</td>\n",
       "      <td>HELZ2;FDPS;NFYA;CARM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>746</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Formation Of Senescence-Associated Heterochrom...</td>\n",
       "      <td>3/17</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.010383</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>48.445617</td>\n",
       "      <td>471.627237</td>\n",
       "      <td>HIRA;HMGA1;CABIN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Regulation Of Cholesterol Biosynthesis By SREB...</td>\n",
       "      <td>4/55</td>\n",
       "      <td>0.000115</td>\n",
       "      <td>0.013401</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17.902186</td>\n",
       "      <td>162.455222</td>\n",
       "      <td>HELZ2;FDPS;NFYA;CARM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>Panther_2016</td>\n",
       "      <td>TCA cycle Homo sapiens P00051</td>\n",
       "      <td>2/9</td>\n",
       "      <td>0.000722</td>\n",
       "      <td>0.014439</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63.890851</td>\n",
       "      <td>462.157096</td>\n",
       "      <td>CS;ACO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>WikiPathway_2021_Human</td>\n",
       "      <td>Fluoroacetic acid toxicity WP4966</td>\n",
       "      <td>2/6</td>\n",
       "      <td>0.000304</td>\n",
       "      <td>0.029743</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>111.825843</td>\n",
       "      <td>905.802397</td>\n",
       "      <td>CS;ACO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>Elsevier_Pathway_Collection</td>\n",
       "      <td>Myeloid Cell Differentiation Block in Leukemia</td>\n",
       "      <td>3/25</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.035669</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30.816632</td>\n",
       "      <td>263.265573</td>\n",
       "      <td>CSF1R;SPI1;FLT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>Glyoxylate and dicarboxylate metabolism</td>\n",
       "      <td>3/30</td>\n",
       "      <td>0.000338</td>\n",
       "      <td>0.036890</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25.103535</td>\n",
       "      <td>200.606096</td>\n",
       "      <td>CS;AMT;ACO2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Gene_set  \\\n",
       "745                Reactome_2022   \n",
       "746                Reactome_2022   \n",
       "747                Reactome_2022   \n",
       "725                 Panther_2016   \n",
       "335       WikiPathway_2021_Human   \n",
       "433  Elsevier_Pathway_Collection   \n",
       "616              KEGG_2021_Human   \n",
       "\n",
       "                                                  Term Overlap   P-value  \\\n",
       "745  Activation Of Gene Expression By SREBF (SREBP)...    4/42  0.000039   \n",
       "746  Formation Of Senescence-Associated Heterochrom...    3/17  0.000059   \n",
       "747  Regulation Of Cholesterol Biosynthesis By SREB...    4/55  0.000115   \n",
       "725                      TCA cycle Homo sapiens P00051     2/9  0.000722   \n",
       "335                  Fluoroacetic acid toxicity WP4966     2/6  0.000304   \n",
       "433     Myeloid Cell Differentiation Block in Leukemia    3/25  0.000195   \n",
       "616            Glyoxylate and dicarboxylate metabolism    3/30  0.000338   \n",
       "\n",
       "     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
       "745          0.010383            0                     0   24.042347   \n",
       "746          0.010383            0                     0   48.445617   \n",
       "747          0.013401            0                     0   17.902186   \n",
       "725          0.014439            0                     0   63.890851   \n",
       "335          0.029743            0                     0  111.825843   \n",
       "433          0.035669            0                     0   30.816632   \n",
       "616          0.036890            0                     0   25.103535   \n",
       "\n",
       "     Combined Score                  Genes  \n",
       "745      243.877579  HELZ2;FDPS;NFYA;CARM1  \n",
       "746      471.627237      HIRA;HMGA1;CABIN1  \n",
       "747      162.455222  HELZ2;FDPS;NFYA;CARM1  \n",
       "725      462.157096                CS;ACO2  \n",
       "335      905.802397                CS;ACO2  \n",
       "433      263.265573        CSF1R;SPI1;FLT3  \n",
       "616      200.606096            CS;AMT;ACO2  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We now perform over-representation analysis with over-expressed genes\n",
    "\n",
    "enr_up = gp.enrich(gene_list=gene_up, \n",
    "                gene_sets=['BioCarta_2016', 'BioPlanet_2019', 'WikiPathway_2021_Human',\\\n",
    "                           'Elsevier_Pathway_Collection', 'KEGG_2021_Human', 'Panther_2016', 'Reactome_2022',\\\n",
    "                           'GO_Biological_Process_2021'], \n",
    "                background=df_tot.columns)\n",
    "\n",
    "enr_up.results[enr_up.results['Adjusted P-value']<alpha].sort_values(by=['Adjusted P-value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fc4292d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We plot the best results of over-representation analysis on over-expressed genes \n",
    "\n",
    "ax = dotplot(enr_up.results,\n",
    "              column=\"Adjusted P-value\",\n",
    "              x='Gene_set', # set x axis, so you could do a multi-sample/library comparsion\n",
    "              size=10,\n",
    "              top_term=3,\n",
    "              figsize=(3,5),\n",
    "              title = \"OVER EXPRESSED\",\n",
    "              xticklabels_rot=45, # rotate xtick labels\n",
    "              show_ring=True, # set to False to revmove outer ring\n",
    "              marker='o',\n",
    "              ofname='up_genes_dotlot.png'\n",
    "             )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca1815fb",
   "metadata": {},
   "source": [
    "### All differentially expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "94f27cdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1839</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Metabolism Of RNA R-HSA-8953854</td>\n",
       "      <td>58/666</td>\n",
       "      <td>8.437535e-21</td>\n",
       "      <td>7.534719e-18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.129426</td>\n",
       "      <td>237.090276</td>\n",
       "      <td>EIF4A2;POP5;HBS1L;RPL30;RPL31;SMG7;CWC27;RPL7;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1840</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Translation R-HSA-72766</td>\n",
       "      <td>36/281</td>\n",
       "      <td>1.525355e-18</td>\n",
       "      <td>6.810709e-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7.579690</td>\n",
       "      <td>310.951503</td>\n",
       "      <td>EIF4A2;SRP19;RPL30;RPL31;MRPS14;RPL36A;MRPL17;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>BioPlanet_2019</td>\n",
       "      <td>Translation</td>\n",
       "      <td>27/151</td>\n",
       "      <td>6.108237e-18</td>\n",
       "      <td>4.257441e-15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11.037361</td>\n",
       "      <td>437.486718</td>\n",
       "      <td>EIF4A2;SRP19;RPL30;RPL31;RPL36A;SRP14;RPL7;RPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2732</th>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "      <td>cytoplasmic translation (GO:0002181)</td>\n",
       "      <td>22/93</td>\n",
       "      <td>1.383530e-17</td>\n",
       "      <td>1.985341e-14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15.547123</td>\n",
       "      <td>603.528581</td>\n",
       "      <td>EIF4A2;RPL30;RPL41;RPL21;RPS7;RPL31;RPL23;RPL3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2733</th>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "      <td>cotranslational protein targeting to membrane ...</td>\n",
       "      <td>22/94</td>\n",
       "      <td>1.772626e-17</td>\n",
       "      <td>1.985341e-14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15.330406</td>\n",
       "      <td>591.316505</td>\n",
       "      <td>SRP19;RPL30;RPL41;RPL21;RPS7;RPL31;RPL23;RPL36...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1952</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Transport Of SLBP Dependant Mature mRNA R-HSA-...</td>\n",
       "      <td>4/36</td>\n",
       "      <td>6.157601e-03</td>\n",
       "      <td>4.823454e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.006450</td>\n",
       "      <td>30.573238</td>\n",
       "      <td>SLBP;NUP35;EIF4E;NUP37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2768</th>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "      <td>regulation of ubiquitin protein ligase activit...</td>\n",
       "      <td>4/21</td>\n",
       "      <td>7.970941e-04</td>\n",
       "      <td>4.825651e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11.314930</td>\n",
       "      <td>80.726795</td>\n",
       "      <td>RAB1A;RPS7;RPL23;SKP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Cytoprotection By HMOX1 R-HSA-9707564</td>\n",
       "      <td>5/57</td>\n",
       "      <td>6.236425e-03</td>\n",
       "      <td>4.842719e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.626989</td>\n",
       "      <td>23.492836</td>\n",
       "      <td>HELZ2;COX16;COX4I1;CARM1;COX7C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>BioPlanet_2019</td>\n",
       "      <td>Messenger RNA splicing: major pathway</td>\n",
       "      <td>6/68</td>\n",
       "      <td>2.716345e-03</td>\n",
       "      <td>4.854597e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.665950</td>\n",
       "      <td>27.568615</td>\n",
       "      <td>PABPN1;SF3B6;HNRNPA2B1;PAPOLA;DNAJC8;HNRNPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1954</th>\n",
       "      <td>Reactome_2022</td>\n",
       "      <td>Signaling By NOTCH4 R-HSA-9013694</td>\n",
       "      <td>6/81</td>\n",
       "      <td>6.451424e-03</td>\n",
       "      <td>4.966484e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.854617</td>\n",
       "      <td>19.440587</td>\n",
       "      <td>PSMA3;NCSTN;PSMC6;PSMD14;PSMA2;SKP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Gene_set  \\\n",
       "1839               Reactome_2022   \n",
       "1840               Reactome_2022   \n",
       "67                BioPlanet_2019   \n",
       "2732  GO_Biological_Process_2021   \n",
       "2733  GO_Biological_Process_2021   \n",
       "...                          ...   \n",
       "1952               Reactome_2022   \n",
       "2768  GO_Biological_Process_2021   \n",
       "1953               Reactome_2022   \n",
       "105               BioPlanet_2019   \n",
       "1954               Reactome_2022   \n",
       "\n",
       "                                                   Term Overlap       P-value  \\\n",
       "1839                    Metabolism Of RNA R-HSA-8953854  58/666  8.437535e-21   \n",
       "1840                            Translation R-HSA-72766  36/281  1.525355e-18   \n",
       "67                                          Translation  27/151  6.108237e-18   \n",
       "2732               cytoplasmic translation (GO:0002181)   22/93  1.383530e-17   \n",
       "2733  cotranslational protein targeting to membrane ...   22/94  1.772626e-17   \n",
       "...                                                 ...     ...           ...   \n",
       "1952  Transport Of SLBP Dependant Mature mRNA R-HSA-...    4/36  6.157601e-03   \n",
       "2768  regulation of ubiquitin protein ligase activit...    4/21  7.970941e-04   \n",
       "1953              Cytoprotection By HMOX1 R-HSA-9707564    5/57  6.236425e-03   \n",
       "105               Messenger RNA splicing: major pathway    6/68  2.716345e-03   \n",
       "1954                  Signaling By NOTCH4 R-HSA-9013694    6/81  6.451424e-03   \n",
       "\n",
       "      Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
       "1839      7.534719e-18            0                     0    5.129426   \n",
       "1840      6.810709e-16            0                     0    7.579690   \n",
       "67        4.257441e-15            0                     0   11.037361   \n",
       "2732      1.985341e-14            0                     0   15.547123   \n",
       "2733      1.985341e-14            0                     0   15.330406   \n",
       "...                ...          ...                   ...         ...   \n",
       "1952      4.823454e-02            0                     0    6.006450   \n",
       "2768      4.825651e-02            0                     0   11.314930   \n",
       "1953      4.842719e-02            0                     0    4.626989   \n",
       "105       4.854597e-02            0                     0    4.665950   \n",
       "1954      4.966484e-02            0                     0    3.854617   \n",
       "\n",
       "      Combined Score                                              Genes  \n",
       "1839      237.090276  EIF4A2;POP5;HBS1L;RPL30;RPL31;SMG7;CWC27;RPL7;...  \n",
       "1840      310.951503  EIF4A2;SRP19;RPL30;RPL31;MRPS14;RPL36A;MRPL17;...  \n",
       "67        437.486718  EIF4A2;SRP19;RPL30;RPL31;RPL36A;SRP14;RPL7;RPS...  \n",
       "2732      603.528581  EIF4A2;RPL30;RPL41;RPL21;RPS7;RPL31;RPL23;RPL3...  \n",
       "2733      591.316505  SRP19;RPL30;RPL41;RPL21;RPS7;RPL31;RPL23;RPL36...  \n",
       "...              ...                                                ...  \n",
       "1952       30.573238                             SLBP;NUP35;EIF4E;NUP37  \n",
       "2768       80.726795                              RAB1A;RPS7;RPL23;SKP1  \n",
       "1953       23.492836                     HELZ2;COX16;COX4I1;CARM1;COX7C  \n",
       "105        27.568615        PABPN1;SF3B6;HNRNPA2B1;PAPOLA;DNAJC8;HNRNPC  \n",
       "1954       19.440587                PSMA3;NCSTN;PSMC6;PSMD14;PSMA2;SKP1  \n",
       "\n",
       "[213 rows x 10 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enr_tot = gp.enrich(gene_list=list(diff_gen), \n",
    "                gene_sets=['BioCarta_2016', 'BioPlanet_2019', 'WikiPathway_2021_Human',\\\n",
    "                           'Elsevier_Pathway_Collection', 'KEGG_2021_Human', 'Panther_2016', 'Reactome_2022',\\\n",
    "                           'GO_Biological_Process_2021'], \n",
    "                background=df_tot.columns)\n",
    "\n",
    "enr_tot.results[enr_tot.results['Adjusted P-value']<alpha].sort_values(by=['Adjusted P-value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "241e6f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We plot results for all genes together\n",
    "\n",
    "ax = dotplot(enr_tot.results,\n",
    "              column=\"Adjusted P-value\",\n",
    "              x='Gene_set', # set x axis, so you could do a multi-sample/library comparsion\n",
    "              size=10,\n",
    "              top_term=3,\n",
    "              figsize=(3,5),\n",
    "              title = \"ALL GENES\",\n",
    "              xticklabels_rot=45, # rotate xtick labels\n",
    "              show_ring=True, # set to False to revmove outer ring\n",
    "              marker='o',\n",
    "              ofname='all_genes_dotlot.png'\n",
    "             )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "760a5b25",
   "metadata": {},
   "source": [
    "## Gene Set Enrichment Analysis (GSEA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "13d177d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ZKSCAN8</th>\n",
       "      <th>ZNF532</th>\n",
       "      <th>UBE2I</th>\n",
       "      <th>ANAPC7</th>\n",
       "      <th>CAMK1G</th>\n",
       "      <th>EID3</th>\n",
       "      <th>EMC6</th>\n",
       "      <th>ZMYND12</th>\n",
       "      <th>APOBEC3F</th>\n",
       "      <th>EREG</th>\n",
       "      <th>...</th>\n",
       "      <th>RP1L1</th>\n",
       "      <th>ZNF678</th>\n",
       "      <th>SCHIP1</th>\n",
       "      <th>OR2T1</th>\n",
       "      <th>SH2D4B</th>\n",
       "      <th>LILRB2</th>\n",
       "      <th>FAM170A</th>\n",
       "      <th>CUX2</th>\n",
       "      <th>LIN9</th>\n",
       "      <th>MAP6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM290792</th>\n",
       "      <td>7.836896</td>\n",
       "      <td>11.556448</td>\n",
       "      <td>13.296613</td>\n",
       "      <td>9.808581</td>\n",
       "      <td>8.393730</td>\n",
       "      <td>9.142067</td>\n",
       "      <td>12.965830</td>\n",
       "      <td>8.936077</td>\n",
       "      <td>10.720192</td>\n",
       "      <td>9.983204</td>\n",
       "      <td>...</td>\n",
       "      <td>7.426869</td>\n",
       "      <td>9.758695</td>\n",
       "      <td>7.745713</td>\n",
       "      <td>8.150235</td>\n",
       "      <td>8.165984</td>\n",
       "      <td>18.771669</td>\n",
       "      <td>7.331735</td>\n",
       "      <td>9.821140</td>\n",
       "      <td>9.561976</td>\n",
       "      <td>10.520180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290793</th>\n",
       "      <td>8.454090</td>\n",
       "      <td>9.939125</td>\n",
       "      <td>13.804273</td>\n",
       "      <td>10.497013</td>\n",
       "      <td>7.523967</td>\n",
       "      <td>9.362625</td>\n",
       "      <td>13.239098</td>\n",
       "      <td>9.572343</td>\n",
       "      <td>10.886928</td>\n",
       "      <td>8.181475</td>\n",
       "      <td>...</td>\n",
       "      <td>7.885363</td>\n",
       "      <td>8.961994</td>\n",
       "      <td>8.769891</td>\n",
       "      <td>8.552343</td>\n",
       "      <td>7.420828</td>\n",
       "      <td>18.545509</td>\n",
       "      <td>7.663930</td>\n",
       "      <td>7.601909</td>\n",
       "      <td>8.743873</td>\n",
       "      <td>9.655440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290794</th>\n",
       "      <td>7.278090</td>\n",
       "      <td>9.881257</td>\n",
       "      <td>13.383527</td>\n",
       "      <td>9.261186</td>\n",
       "      <td>8.101607</td>\n",
       "      <td>8.923248</td>\n",
       "      <td>13.074373</td>\n",
       "      <td>7.928118</td>\n",
       "      <td>9.894150</td>\n",
       "      <td>7.895988</td>\n",
       "      <td>...</td>\n",
       "      <td>8.051473</td>\n",
       "      <td>9.812328</td>\n",
       "      <td>8.709778</td>\n",
       "      <td>8.041324</td>\n",
       "      <td>7.699717</td>\n",
       "      <td>19.076270</td>\n",
       "      <td>8.174980</td>\n",
       "      <td>9.359250</td>\n",
       "      <td>7.813959</td>\n",
       "      <td>10.128955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290795</th>\n",
       "      <td>10.258611</td>\n",
       "      <td>10.711247</td>\n",
       "      <td>13.520033</td>\n",
       "      <td>10.150032</td>\n",
       "      <td>6.122135</td>\n",
       "      <td>9.998456</td>\n",
       "      <td>13.016399</td>\n",
       "      <td>7.187788</td>\n",
       "      <td>10.841216</td>\n",
       "      <td>8.695254</td>\n",
       "      <td>...</td>\n",
       "      <td>8.666089</td>\n",
       "      <td>11.738657</td>\n",
       "      <td>8.396363</td>\n",
       "      <td>8.904460</td>\n",
       "      <td>7.766847</td>\n",
       "      <td>18.194124</td>\n",
       "      <td>8.911766</td>\n",
       "      <td>9.294017</td>\n",
       "      <td>7.819357</td>\n",
       "      <td>9.621316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290796</th>\n",
       "      <td>7.134759</td>\n",
       "      <td>10.251259</td>\n",
       "      <td>13.151941</td>\n",
       "      <td>8.837736</td>\n",
       "      <td>8.454090</td>\n",
       "      <td>8.966248</td>\n",
       "      <td>13.766642</td>\n",
       "      <td>8.211519</td>\n",
       "      <td>10.357897</td>\n",
       "      <td>8.672921</td>\n",
       "      <td>...</td>\n",
       "      <td>7.790092</td>\n",
       "      <td>11.562495</td>\n",
       "      <td>7.367434</td>\n",
       "      <td>8.760854</td>\n",
       "      <td>7.733277</td>\n",
       "      <td>17.982601</td>\n",
       "      <td>10.218124</td>\n",
       "      <td>9.800906</td>\n",
       "      <td>8.391254</td>\n",
       "      <td>10.847043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290797</th>\n",
       "      <td>6.972344</td>\n",
       "      <td>11.385409</td>\n",
       "      <td>13.952971</td>\n",
       "      <td>9.372846</td>\n",
       "      <td>8.626830</td>\n",
       "      <td>9.633524</td>\n",
       "      <td>13.044780</td>\n",
       "      <td>8.346950</td>\n",
       "      <td>8.543487</td>\n",
       "      <td>8.645184</td>\n",
       "      <td>...</td>\n",
       "      <td>7.947785</td>\n",
       "      <td>10.729930</td>\n",
       "      <td>8.419690</td>\n",
       "      <td>7.562210</td>\n",
       "      <td>8.829405</td>\n",
       "      <td>17.888218</td>\n",
       "      <td>8.003113</td>\n",
       "      <td>9.642482</td>\n",
       "      <td>8.859554</td>\n",
       "      <td>10.529190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290798</th>\n",
       "      <td>8.021632</td>\n",
       "      <td>11.025483</td>\n",
       "      <td>14.027703</td>\n",
       "      <td>10.217827</td>\n",
       "      <td>8.449017</td>\n",
       "      <td>8.911766</td>\n",
       "      <td>13.172426</td>\n",
       "      <td>7.812677</td>\n",
       "      <td>8.696229</td>\n",
       "      <td>8.896148</td>\n",
       "      <td>...</td>\n",
       "      <td>8.743675</td>\n",
       "      <td>10.147113</td>\n",
       "      <td>8.382042</td>\n",
       "      <td>8.969826</td>\n",
       "      <td>8.226455</td>\n",
       "      <td>18.100123</td>\n",
       "      <td>7.694913</td>\n",
       "      <td>8.780831</td>\n",
       "      <td>7.808374</td>\n",
       "      <td>10.488346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290799</th>\n",
       "      <td>8.680602</td>\n",
       "      <td>9.983204</td>\n",
       "      <td>13.500088</td>\n",
       "      <td>9.915698</td>\n",
       "      <td>8.972284</td>\n",
       "      <td>9.030126</td>\n",
       "      <td>13.030929</td>\n",
       "      <td>8.346127</td>\n",
       "      <td>10.423708</td>\n",
       "      <td>9.205629</td>\n",
       "      <td>...</td>\n",
       "      <td>8.449144</td>\n",
       "      <td>10.461621</td>\n",
       "      <td>8.427326</td>\n",
       "      <td>8.588333</td>\n",
       "      <td>7.615233</td>\n",
       "      <td>17.923042</td>\n",
       "      <td>7.887946</td>\n",
       "      <td>9.362625</td>\n",
       "      <td>9.119511</td>\n",
       "      <td>10.357344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290800</th>\n",
       "      <td>8.281709</td>\n",
       "      <td>11.062754</td>\n",
       "      <td>13.511526</td>\n",
       "      <td>9.133993</td>\n",
       "      <td>7.830803</td>\n",
       "      <td>9.405434</td>\n",
       "      <td>12.924767</td>\n",
       "      <td>8.198483</td>\n",
       "      <td>10.134095</td>\n",
       "      <td>9.619962</td>\n",
       "      <td>...</td>\n",
       "      <td>7.823325</td>\n",
       "      <td>9.908717</td>\n",
       "      <td>7.469684</td>\n",
       "      <td>9.132324</td>\n",
       "      <td>9.103808</td>\n",
       "      <td>18.325386</td>\n",
       "      <td>7.419842</td>\n",
       "      <td>10.236709</td>\n",
       "      <td>8.658698</td>\n",
       "      <td>10.634424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290807</th>\n",
       "      <td>7.531574</td>\n",
       "      <td>9.735370</td>\n",
       "      <td>13.945415</td>\n",
       "      <td>10.044735</td>\n",
       "      <td>7.964978</td>\n",
       "      <td>9.830841</td>\n",
       "      <td>12.879289</td>\n",
       "      <td>8.155777</td>\n",
       "      <td>10.018769</td>\n",
       "      <td>9.216510</td>\n",
       "      <td>...</td>\n",
       "      <td>8.270849</td>\n",
       "      <td>10.547232</td>\n",
       "      <td>8.374836</td>\n",
       "      <td>9.283694</td>\n",
       "      <td>7.505329</td>\n",
       "      <td>18.756286</td>\n",
       "      <td>8.211897</td>\n",
       "      <td>10.056449</td>\n",
       "      <td>8.428890</td>\n",
       "      <td>10.495666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290812</th>\n",
       "      <td>7.337679</td>\n",
       "      <td>10.544480</td>\n",
       "      <td>13.553340</td>\n",
       "      <td>10.062991</td>\n",
       "      <td>7.244023</td>\n",
       "      <td>8.996607</td>\n",
       "      <td>13.202428</td>\n",
       "      <td>8.933978</td>\n",
       "      <td>9.240168</td>\n",
       "      <td>9.249096</td>\n",
       "      <td>...</td>\n",
       "      <td>7.044738</td>\n",
       "      <td>11.593240</td>\n",
       "      <td>7.879427</td>\n",
       "      <td>7.357871</td>\n",
       "      <td>8.088103</td>\n",
       "      <td>17.960436</td>\n",
       "      <td>8.109462</td>\n",
       "      <td>10.522151</td>\n",
       "      <td>8.211362</td>\n",
       "      <td>10.132095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290816</th>\n",
       "      <td>8.711001</td>\n",
       "      <td>11.998301</td>\n",
       "      <td>12.888501</td>\n",
       "      <td>8.875158</td>\n",
       "      <td>8.083784</td>\n",
       "      <td>9.526388</td>\n",
       "      <td>13.524342</td>\n",
       "      <td>7.816948</td>\n",
       "      <td>9.270313</td>\n",
       "      <td>8.178023</td>\n",
       "      <td>...</td>\n",
       "      <td>9.476331</td>\n",
       "      <td>10.594055</td>\n",
       "      <td>7.819889</td>\n",
       "      <td>7.079614</td>\n",
       "      <td>7.686465</td>\n",
       "      <td>18.273273</td>\n",
       "      <td>8.075330</td>\n",
       "      <td>8.785462</td>\n",
       "      <td>7.615529</td>\n",
       "      <td>10.737178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290817</th>\n",
       "      <td>7.370765</td>\n",
       "      <td>10.746211</td>\n",
       "      <td>13.683590</td>\n",
       "      <td>10.629111</td>\n",
       "      <td>8.047821</td>\n",
       "      <td>9.767217</td>\n",
       "      <td>13.734776</td>\n",
       "      <td>8.835665</td>\n",
       "      <td>11.061831</td>\n",
       "      <td>8.031646</td>\n",
       "      <td>...</td>\n",
       "      <td>9.203081</td>\n",
       "      <td>9.550889</td>\n",
       "      <td>8.393537</td>\n",
       "      <td>8.772018</td>\n",
       "      <td>8.044503</td>\n",
       "      <td>17.291196</td>\n",
       "      <td>7.669113</td>\n",
       "      <td>7.152983</td>\n",
       "      <td>9.332863</td>\n",
       "      <td>9.639253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290818</th>\n",
       "      <td>7.616559</td>\n",
       "      <td>11.958278</td>\n",
       "      <td>12.248943</td>\n",
       "      <td>7.992173</td>\n",
       "      <td>7.399937</td>\n",
       "      <td>9.875525</td>\n",
       "      <td>13.473040</td>\n",
       "      <td>8.325490</td>\n",
       "      <td>9.554910</td>\n",
       "      <td>7.872808</td>\n",
       "      <td>...</td>\n",
       "      <td>8.219204</td>\n",
       "      <td>11.007826</td>\n",
       "      <td>7.302278</td>\n",
       "      <td>9.167173</td>\n",
       "      <td>8.558747</td>\n",
       "      <td>18.316886</td>\n",
       "      <td>7.481784</td>\n",
       "      <td>9.335869</td>\n",
       "      <td>9.288055</td>\n",
       "      <td>10.567470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290819</th>\n",
       "      <td>7.724920</td>\n",
       "      <td>9.851807</td>\n",
       "      <td>14.121267</td>\n",
       "      <td>9.555641</td>\n",
       "      <td>7.812956</td>\n",
       "      <td>10.055975</td>\n",
       "      <td>13.521172</td>\n",
       "      <td>7.852051</td>\n",
       "      <td>9.804600</td>\n",
       "      <td>7.580368</td>\n",
       "      <td>...</td>\n",
       "      <td>7.706708</td>\n",
       "      <td>10.136614</td>\n",
       "      <td>7.871405</td>\n",
       "      <td>7.788657</td>\n",
       "      <td>7.608933</td>\n",
       "      <td>17.940482</td>\n",
       "      <td>7.675535</td>\n",
       "      <td>10.409149</td>\n",
       "      <td>9.102335</td>\n",
       "      <td>9.191861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290820</th>\n",
       "      <td>7.695143</td>\n",
       "      <td>10.665347</td>\n",
       "      <td>13.814970</td>\n",
       "      <td>9.456169</td>\n",
       "      <td>8.942591</td>\n",
       "      <td>10.196694</td>\n",
       "      <td>13.256555</td>\n",
       "      <td>8.265839</td>\n",
       "      <td>9.913469</td>\n",
       "      <td>7.951397</td>\n",
       "      <td>...</td>\n",
       "      <td>7.793898</td>\n",
       "      <td>10.148575</td>\n",
       "      <td>6.087737</td>\n",
       "      <td>7.628597</td>\n",
       "      <td>8.781164</td>\n",
       "      <td>17.966733</td>\n",
       "      <td>6.814486</td>\n",
       "      <td>9.216344</td>\n",
       "      <td>8.894234</td>\n",
       "      <td>9.433552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290821</th>\n",
       "      <td>8.801931</td>\n",
       "      <td>10.104682</td>\n",
       "      <td>13.650664</td>\n",
       "      <td>10.642495</td>\n",
       "      <td>7.174347</td>\n",
       "      <td>8.443850</td>\n",
       "      <td>13.360218</td>\n",
       "      <td>8.249108</td>\n",
       "      <td>8.759114</td>\n",
       "      <td>8.157565</td>\n",
       "      <td>...</td>\n",
       "      <td>7.210114</td>\n",
       "      <td>8.953207</td>\n",
       "      <td>8.751703</td>\n",
       "      <td>9.764340</td>\n",
       "      <td>8.704036</td>\n",
       "      <td>18.716765</td>\n",
       "      <td>6.923515</td>\n",
       "      <td>9.058786</td>\n",
       "      <td>7.865130</td>\n",
       "      <td>9.838539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290822</th>\n",
       "      <td>9.520554</td>\n",
       "      <td>9.866795</td>\n",
       "      <td>13.529753</td>\n",
       "      <td>9.769843</td>\n",
       "      <td>9.204245</td>\n",
       "      <td>9.119734</td>\n",
       "      <td>13.631693</td>\n",
       "      <td>9.086685</td>\n",
       "      <td>9.604065</td>\n",
       "      <td>7.946977</td>\n",
       "      <td>...</td>\n",
       "      <td>7.313767</td>\n",
       "      <td>11.490777</td>\n",
       "      <td>7.175653</td>\n",
       "      <td>8.437295</td>\n",
       "      <td>8.517987</td>\n",
       "      <td>17.988050</td>\n",
       "      <td>9.520988</td>\n",
       "      <td>9.673355</td>\n",
       "      <td>9.168202</td>\n",
       "      <td>9.778815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290823</th>\n",
       "      <td>7.756522</td>\n",
       "      <td>9.688699</td>\n",
       "      <td>14.372324</td>\n",
       "      <td>10.985775</td>\n",
       "      <td>7.876163</td>\n",
       "      <td>9.606982</td>\n",
       "      <td>13.314560</td>\n",
       "      <td>8.079909</td>\n",
       "      <td>10.569335</td>\n",
       "      <td>7.643912</td>\n",
       "      <td>...</td>\n",
       "      <td>8.588011</td>\n",
       "      <td>9.359134</td>\n",
       "      <td>7.824058</td>\n",
       "      <td>7.820237</td>\n",
       "      <td>8.148928</td>\n",
       "      <td>18.273273</td>\n",
       "      <td>7.388638</td>\n",
       "      <td>8.866602</td>\n",
       "      <td>8.882471</td>\n",
       "      <td>8.290506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM290824</th>\n",
       "      <td>8.652816</td>\n",
       "      <td>10.023413</td>\n",
       "      <td>14.179916</td>\n",
       "      <td>10.127081</td>\n",
       "      <td>7.241005</td>\n",
       "      <td>9.577288</td>\n",
       "      <td>13.731312</td>\n",
       "      <td>7.512735</td>\n",
       "      <td>10.023994</td>\n",
       "      <td>7.339665</td>\n",
       "      <td>...</td>\n",
       "      <td>8.417735</td>\n",
       "      <td>9.828095</td>\n",
       "      <td>7.787604</td>\n",
       "      <td>7.107691</td>\n",
       "      <td>8.091075</td>\n",
       "      <td>17.609236</td>\n",
       "      <td>7.759362</td>\n",
       "      <td>8.881503</td>\n",
       "      <td>8.503289</td>\n",
       "      <td>9.700231</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 17916 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ZKSCAN8     ZNF532      UBE2I     ANAPC7    CAMK1G       EID3  \\\n",
       "GSM290792   7.836896  11.556448  13.296613   9.808581  8.393730   9.142067   \n",
       "GSM290793   8.454090   9.939125  13.804273  10.497013  7.523967   9.362625   \n",
       "GSM290794   7.278090   9.881257  13.383527   9.261186  8.101607   8.923248   \n",
       "GSM290795  10.258611  10.711247  13.520033  10.150032  6.122135   9.998456   \n",
       "GSM290796   7.134759  10.251259  13.151941   8.837736  8.454090   8.966248   \n",
       "GSM290797   6.972344  11.385409  13.952971   9.372846  8.626830   9.633524   \n",
       "GSM290798   8.021632  11.025483  14.027703  10.217827  8.449017   8.911766   \n",
       "GSM290799   8.680602   9.983204  13.500088   9.915698  8.972284   9.030126   \n",
       "GSM290800   8.281709  11.062754  13.511526   9.133993  7.830803   9.405434   \n",
       "GSM290807   7.531574   9.735370  13.945415  10.044735  7.964978   9.830841   \n",
       "GSM290812   7.337679  10.544480  13.553340  10.062991  7.244023   8.996607   \n",
       "GSM290816   8.711001  11.998301  12.888501   8.875158  8.083784   9.526388   \n",
       "GSM290817   7.370765  10.746211  13.683590  10.629111  8.047821   9.767217   \n",
       "GSM290818   7.616559  11.958278  12.248943   7.992173  7.399937   9.875525   \n",
       "GSM290819   7.724920   9.851807  14.121267   9.555641  7.812956  10.055975   \n",
       "GSM290820   7.695143  10.665347  13.814970   9.456169  8.942591  10.196694   \n",
       "GSM290821   8.801931  10.104682  13.650664  10.642495  7.174347   8.443850   \n",
       "GSM290822   9.520554   9.866795  13.529753   9.769843  9.204245   9.119734   \n",
       "GSM290823   7.756522   9.688699  14.372324  10.985775  7.876163   9.606982   \n",
       "GSM290824   8.652816  10.023413  14.179916  10.127081  7.241005   9.577288   \n",
       "\n",
       "                EMC6   ZMYND12   APOBEC3F      EREG  ...     RP1L1     ZNF678  \\\n",
       "GSM290792  12.965830  8.936077  10.720192  9.983204  ...  7.426869   9.758695   \n",
       "GSM290793  13.239098  9.572343  10.886928  8.181475  ...  7.885363   8.961994   \n",
       "GSM290794  13.074373  7.928118   9.894150  7.895988  ...  8.051473   9.812328   \n",
       "GSM290795  13.016399  7.187788  10.841216  8.695254  ...  8.666089  11.738657   \n",
       "GSM290796  13.766642  8.211519  10.357897  8.672921  ...  7.790092  11.562495   \n",
       "GSM290797  13.044780  8.346950   8.543487  8.645184  ...  7.947785  10.729930   \n",
       "GSM290798  13.172426  7.812677   8.696229  8.896148  ...  8.743675  10.147113   \n",
       "GSM290799  13.030929  8.346127  10.423708  9.205629  ...  8.449144  10.461621   \n",
       "GSM290800  12.924767  8.198483  10.134095  9.619962  ...  7.823325   9.908717   \n",
       "GSM290807  12.879289  8.155777  10.018769  9.216510  ...  8.270849  10.547232   \n",
       "GSM290812  13.202428  8.933978   9.240168  9.249096  ...  7.044738  11.593240   \n",
       "GSM290816  13.524342  7.816948   9.270313  8.178023  ...  9.476331  10.594055   \n",
       "GSM290817  13.734776  8.835665  11.061831  8.031646  ...  9.203081   9.550889   \n",
       "GSM290818  13.473040  8.325490   9.554910  7.872808  ...  8.219204  11.007826   \n",
       "GSM290819  13.521172  7.852051   9.804600  7.580368  ...  7.706708  10.136614   \n",
       "GSM290820  13.256555  8.265839   9.913469  7.951397  ...  7.793898  10.148575   \n",
       "GSM290821  13.360218  8.249108   8.759114  8.157565  ...  7.210114   8.953207   \n",
       "GSM290822  13.631693  9.086685   9.604065  7.946977  ...  7.313767  11.490777   \n",
       "GSM290823  13.314560  8.079909  10.569335  7.643912  ...  8.588011   9.359134   \n",
       "GSM290824  13.731312  7.512735  10.023994  7.339665  ...  8.417735   9.828095   \n",
       "\n",
       "             SCHIP1     OR2T1    SH2D4B     LILRB2    FAM170A       CUX2  \\\n",
       "GSM290792  7.745713  8.150235  8.165984  18.771669   7.331735   9.821140   \n",
       "GSM290793  8.769891  8.552343  7.420828  18.545509   7.663930   7.601909   \n",
       "GSM290794  8.709778  8.041324  7.699717  19.076270   8.174980   9.359250   \n",
       "GSM290795  8.396363  8.904460  7.766847  18.194124   8.911766   9.294017   \n",
       "GSM290796  7.367434  8.760854  7.733277  17.982601  10.218124   9.800906   \n",
       "GSM290797  8.419690  7.562210  8.829405  17.888218   8.003113   9.642482   \n",
       "GSM290798  8.382042  8.969826  8.226455  18.100123   7.694913   8.780831   \n",
       "GSM290799  8.427326  8.588333  7.615233  17.923042   7.887946   9.362625   \n",
       "GSM290800  7.469684  9.132324  9.103808  18.325386   7.419842  10.236709   \n",
       "GSM290807  8.374836  9.283694  7.505329  18.756286   8.211897  10.056449   \n",
       "GSM290812  7.879427  7.357871  8.088103  17.960436   8.109462  10.522151   \n",
       "GSM290816  7.819889  7.079614  7.686465  18.273273   8.075330   8.785462   \n",
       "GSM290817  8.393537  8.772018  8.044503  17.291196   7.669113   7.152983   \n",
       "GSM290818  7.302278  9.167173  8.558747  18.316886   7.481784   9.335869   \n",
       "GSM290819  7.871405  7.788657  7.608933  17.940482   7.675535  10.409149   \n",
       "GSM290820  6.087737  7.628597  8.781164  17.966733   6.814486   9.216344   \n",
       "GSM290821  8.751703  9.764340  8.704036  18.716765   6.923515   9.058786   \n",
       "GSM290822  7.175653  8.437295  8.517987  17.988050   9.520988   9.673355   \n",
       "GSM290823  7.824058  7.820237  8.148928  18.273273   7.388638   8.866602   \n",
       "GSM290824  7.787604  7.107691  8.091075  17.609236   7.759362   8.881503   \n",
       "\n",
       "               LIN9       MAP6  \n",
       "GSM290792  9.561976  10.520180  \n",
       "GSM290793  8.743873   9.655440  \n",
       "GSM290794  7.813959  10.128955  \n",
       "GSM290795  7.819357   9.621316  \n",
       "GSM290796  8.391254  10.847043  \n",
       "GSM290797  8.859554  10.529190  \n",
       "GSM290798  7.808374  10.488346  \n",
       "GSM290799  9.119511  10.357344  \n",
       "GSM290800  8.658698  10.634424  \n",
       "GSM290807  8.428890  10.495666  \n",
       "GSM290812  8.211362  10.132095  \n",
       "GSM290816  7.615529  10.737178  \n",
       "GSM290817  9.332863   9.639253  \n",
       "GSM290818  9.288055  10.567470  \n",
       "GSM290819  9.102335   9.191861  \n",
       "GSM290820  8.894234   9.433552  \n",
       "GSM290821  7.865130   9.838539  \n",
       "GSM290822  9.168202   9.778815  \n",
       "GSM290823  8.882471   8.290506  \n",
       "GSM290824  8.503289   9.700231  \n",
       "\n",
       "[20 rows x 17916 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#In order to perform GSEA, we define the entire dataset, concatenating ill and healthy patients\n",
    "\n",
    "df_gsea = pd.concat((df_ill, df_health))\n",
    "df_gsea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "432e72a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a cls file for function gsea\n",
    "\n",
    "groups = ['Tumor', 'Tumor', 'Tumor', 'Tumor', 'Tumor', 'Tumor',\\\n",
    "          'Tumor', 'Tumor', 'Tumor', 'Tumor', 'Tumor',' Health',\\\n",
    "          'Health', 'Health', 'Health', 'Health', 'Health', 'Health',\\\n",
    "          'Health', 'Health'] \n",
    "\n",
    "with open('gsea_names.cls', 'w') as cl:\n",
    "    line = f'{len(groups)} 2 1 \\n# Tumor Health\\n' \n",
    "    cl.write(line) \n",
    "    cl.write(' '.join(groups) + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b8e26f9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-01-04 18:17:25,179 [WARNING] Dropping duplicated gene names, only keep the first values\n"
     ]
    }
   ],
   "source": [
    "#We run GSEA using gseapy function gsea\n",
    "\n",
    "gsea = gp.gsea(pd.DataFrame(df_gsea.values.transpose(), index=column_names),\\\n",
    "               gene_sets=['BioCarta_2016', 'BioPlanet_2019', 'WikiPathway_2021_Human',\\\n",
    "                           'Elsevier_Pathway_Collection', 'KEGG_2021_Human', 'Panther_2016', 'Reactome_2022',\\\n",
    "                           'GO_Biological_Process_2021'],\\\n",
    "               cls = 'gsea_names.cls',\\\n",
    "               method='t_test')\n",
    "\n",
    "#REM: we specified to use t-test as a method to rank differentially expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "27363f0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Term</th>\n",
       "      <th>ES</th>\n",
       "      <th>NES</th>\n",
       "      <th>NOM p-val</th>\n",
       "      <th>FDR q-val</th>\n",
       "      <th>FWER p-val</th>\n",
       "      <th>Tag %</th>\n",
       "      <th>Gene %</th>\n",
       "      <th>Lead_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>gsea</td>\n",
       "      <td>BioPlanet_2019__Gap junction assembly</td>\n",
       "      <td>-0.700288</td>\n",
       "      <td>-1.76966</td>\n",
       "      <td>0.003704</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.96</td>\n",
       "      <td>8/17</td>\n",
       "      <td>15.05%</td>\n",
       "      <td>GJD3;GJD4;GJA10;GJB2;GJD2;GJB5;GJB1;GJB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gsea</td>\n",
       "      <td>Elsevier_Pathway_Collection__Gap Junction Asse...</td>\n",
       "      <td>-0.686635</td>\n",
       "      <td>-1.998346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.401835</td>\n",
       "      <td>0.199</td>\n",
       "      <td>9/20</td>\n",
       "      <td>14.56%</td>\n",
       "      <td>GJB2;ADRB1;DRD2;MAPK3;DRD1;MAPK7;EGFR;MAPK1;GJB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>gsea</td>\n",
       "      <td>GO_Biological_Process_2021__focal adhesion ass...</td>\n",
       "      <td>-0.645322</td>\n",
       "      <td>-1.752192</td>\n",
       "      <td>0.004065</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.97</td>\n",
       "      <td>12/19</td>\n",
       "      <td>25.10%</td>\n",
       "      <td>TAOK2;TESK2;TRIP6;SRC;PIP5K1A;BCR;DLC1;ACTN3;R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>gsea</td>\n",
       "      <td>Reactome_2022__FOXO-mediated Transcription Of ...</td>\n",
       "      <td>-0.641668</td>\n",
       "      <td>-1.544623</td>\n",
       "      <td>0.044747</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7/15</td>\n",
       "      <td>12.49%</td>\n",
       "      <td>NFYA;FOXO3;EP300;NFYC;BBC3;CREBBP;CITED2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>gsea</td>\n",
       "      <td>BioPlanet_2019__HDL-mediated lipid transport</td>\n",
       "      <td>-0.641554</td>\n",
       "      <td>-1.612264</td>\n",
       "      <td>0.021442</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7/15</td>\n",
       "      <td>13.73%</td>\n",
       "      <td>CETP;APOE;APOA1;ABCG1;APOC3;APOC2;LCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>gsea</td>\n",
       "      <td>Reactome_2022__Mitochondrial Translation Elong...</td>\n",
       "      <td>0.722428</td>\n",
       "      <td>1.737658</td>\n",
       "      <td>0.007449</td>\n",
       "      <td>0.557684</td>\n",
       "      <td>0.989</td>\n",
       "      <td>53/78</td>\n",
       "      <td>17.38%</td>\n",
       "      <td>MRPL17;MRPL58;MRPS18C;MRPL3;MRPS21;MRPL51;MRPL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>gsea</td>\n",
       "      <td>GO_Biological_Process_2021__cytoplasmic transl...</td>\n",
       "      <td>0.736552</td>\n",
       "      <td>1.650001</td>\n",
       "      <td>0.016129</td>\n",
       "      <td>0.304517</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10/18</td>\n",
       "      <td>15.87%</td>\n",
       "      <td>DENR;EIF4A2;EIF3D;EIF2B3;MCTS1;EIF4B;EIF5;EIF3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>gsea</td>\n",
       "      <td>Reactome_2022__SARS-CoV-2 Modulates Host Trans...</td>\n",
       "      <td>0.738381</td>\n",
       "      <td>1.713796</td>\n",
       "      <td>0.017476</td>\n",
       "      <td>0.413819</td>\n",
       "      <td>0.995</td>\n",
       "      <td>25/43</td>\n",
       "      <td>13.96%</td>\n",
       "      <td>SNRPG;SNRPE;SNRPD2;RPS17;RPS25;RPS3A;RPS20;RPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>gsea</td>\n",
       "      <td>GO_Biological_Process_2021__mitochondrial elec...</td>\n",
       "      <td>0.749547</td>\n",
       "      <td>1.688987</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.377778</td>\n",
       "      <td>0.999</td>\n",
       "      <td>10/16</td>\n",
       "      <td>14.92%</td>\n",
       "      <td>COX7C;COX4I1;COX6C;CYCS;NDUFA4;COX7B;COX7A2L;C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>gsea</td>\n",
       "      <td>WikiPathway_2021_Human__Mitochondrial CIV Asse...</td>\n",
       "      <td>0.770767</td>\n",
       "      <td>1.753893</td>\n",
       "      <td>0.001916</td>\n",
       "      <td>0.574862</td>\n",
       "      <td>0.979</td>\n",
       "      <td>18/26</td>\n",
       "      <td>15.45%</td>\n",
       "      <td>COX7A2;COX16;COX7C;HIGD1A;COX4I1;COX6C;NDUFA4;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6987 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name                                               Term        ES  \\\n",
       "19   gsea              BioPlanet_2019__Gap junction assembly -0.700288   \n",
       "0    gsea  Elsevier_Pathway_Collection__Gap Junction Asse... -0.686635   \n",
       "26   gsea  GO_Biological_Process_2021__focal adhesion ass... -0.645322   \n",
       "392  gsea  Reactome_2022__FOXO-mediated Transcription Of ... -0.641668   \n",
       "221  gsea       BioPlanet_2019__HDL-mediated lipid transport -0.641554   \n",
       "..    ...                                                ...       ...   \n",
       "32   gsea  Reactome_2022__Mitochondrial Translation Elong...  0.722428   \n",
       "139  gsea  GO_Biological_Process_2021__cytoplasmic transl...  0.736552   \n",
       "57   gsea  Reactome_2022__SARS-CoV-2 Modulates Host Trans...  0.738381   \n",
       "77   gsea  GO_Biological_Process_2021__mitochondrial elec...  0.749547   \n",
       "24   gsea  WikiPathway_2021_Human__Mitochondrial CIV Asse...  0.770767   \n",
       "\n",
       "          NES NOM p-val FDR q-val FWER p-val  Tag %  Gene %  \\\n",
       "19   -1.76966  0.003704       1.0       0.96   8/17  15.05%   \n",
       "0   -1.998346       0.0  0.401835      0.199   9/20  14.56%   \n",
       "26  -1.752192  0.004065       1.0       0.97  12/19  25.10%   \n",
       "392 -1.544623  0.044747       1.0        1.0   7/15  12.49%   \n",
       "221 -1.612264  0.021442       1.0        1.0   7/15  13.73%   \n",
       "..        ...       ...       ...        ...    ...     ...   \n",
       "32   1.737658  0.007449  0.557684      0.989  53/78  17.38%   \n",
       "139  1.650001  0.016129  0.304517        1.0  10/18  15.87%   \n",
       "57   1.713796  0.017476  0.413819      0.995  25/43  13.96%   \n",
       "77   1.688987  0.009709  0.377778      0.999  10/16  14.92%   \n",
       "24   1.753893  0.001916  0.574862      0.979  18/26  15.45%   \n",
       "\n",
       "                                            Lead_genes  \n",
       "19            GJD3;GJD4;GJA10;GJB2;GJD2;GJB5;GJB1;GJB4  \n",
       "0     GJB2;ADRB1;DRD2;MAPK3;DRD1;MAPK7;EGFR;MAPK1;GJB1  \n",
       "26   TAOK2;TESK2;TRIP6;SRC;PIP5K1A;BCR;DLC1;ACTN3;R...  \n",
       "392           NFYA;FOXO3;EP300;NFYC;BBC3;CREBBP;CITED2  \n",
       "221             CETP;APOE;APOA1;ABCG1;APOC3;APOC2;LCAT  \n",
       "..                                                 ...  \n",
       "32   MRPL17;MRPL58;MRPS18C;MRPL3;MRPS21;MRPL51;MRPL...  \n",
       "139  DENR;EIF4A2;EIF3D;EIF2B3;MCTS1;EIF4B;EIF5;EIF3...  \n",
       "57   SNRPG;SNRPE;SNRPD2;RPS17;RPS25;RPS3A;RPS20;RPS...  \n",
       "77   COX7C;COX4I1;COX6C;CYCS;NDUFA4;COX7B;COX7A2L;C...  \n",
       "24   COX7A2;COX16;COX7C;HIGD1A;COX4I1;COX6C;NDUFA4;...  \n",
       "\n",
       "[6987 rows x 10 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Show results of GSEA analysis dorte by Enrichment Score\n",
    "\n",
    "gsea.res2d.sort_values(by='ES')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "2dc73d24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Term</th>\n",
       "      <th>ES</th>\n",
       "      <th>NES</th>\n",
       "      <th>NOM p-val</th>\n",
       "      <th>FDR q-val</th>\n",
       "      <th>FWER p-val</th>\n",
       "      <th>Tag %</th>\n",
       "      <th>Gene %</th>\n",
       "      <th>Lead_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>gsea</td>\n",
       "      <td>Reactome_2022__Translation R-HSA-72766</td>\n",
       "      <td>-0.652531</td>\n",
       "      <td>-1.757326</td>\n",
       "      <td>0.007921</td>\n",
       "      <td>0.343112</td>\n",
       "      <td>0.96</td>\n",
       "      <td>152/253</td>\n",
       "      <td>21.85%</td>\n",
       "      <td>RPL26;RPL21;MRPL17;RPS17;RPL41;RPL36A;RPL23;RP...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Name                                    Term        ES       NES  \\\n",
       "43  gsea  Reactome_2022__Translation R-HSA-72766 -0.652531 -1.757326   \n",
       "\n",
       "   NOM p-val FDR q-val FWER p-val    Tag %  Gene %  \\\n",
       "43  0.007921  0.343112       0.96  152/253  21.85%   \n",
       "\n",
       "                                           Lead_genes  \n",
       "43  RPL26;RPL21;MRPL17;RPS17;RPL41;RPL36A;RPL23;RP...  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We look for one pathway we found with previous analysis\n",
    "\n",
    "gsea.res2d[gsea.res2d['Term']=='Reactome_2022__Translation R-HSA-72766']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "6974f44b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We save in a file a plot showing the results of GSEA analysis on the pathway we choose\n",
    "\n",
    "terms = gsea.res2d.Term\n",
    "i = 43  #index of our pathway\n",
    "\n",
    "gp.gseaplot(gsea.ranking, term=terms[i], **gsea.results[terms[i]], ofname='gsea_plot.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "76f80dec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We save in a file an heatmap showing the results of GSEA analysis on the pathway we choose\n",
    "\n",
    "genes = gsea.res2d.Lead_genes[i].split(\";\")\n",
    "\n",
    "ax = heatmap(df = gsea.heatmat.loc[genes], z_score=0, title=terms[i], figsize=(14,4), ofname='gsea_heatmap.png')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
